
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="X-UA-Compatible" content="IE=Edge" />
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
  
  <!-- Licensed under the Apache 2.0 License -->
  <link rel="stylesheet" type="text/css" href="_static/fonts/open-sans/stylesheet.css" />
  <!-- Licensed under the SIL Open Font License -->
  <link rel="stylesheet" type="text/css" href="_static/fonts/source-serif-pro/source-serif-pro.css" />
  <link rel="stylesheet" type="text/css" href="_static/css/bootstrap.min.css" />
  <link rel="stylesheet" type="text/css" href="_static/css/bootstrap-theme.min.css" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
    <title>Example Workflow &#8212; riboraptor 0.3.0 documentation</title>
    <link rel="stylesheet" href="_static/guzzle.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="stylesheet" type="text/css" href="_static/graphviz.css" />
    <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Manual" href="cmd-manual.html" />
    <link rel="prev" title="Installation" href="installation.html" />
  
   

  </head><body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="cmd-manual.html" title="Manual"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="installation.html" title="Installation"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">riboraptor 0.3.0 documentation</a> &#187;</li> 
      </ul>
    </div>
    <div class="container-wrapper">

      <div id="mobile-toggle">
        <a href="#"><span class="glyphicon glyphicon-align-justify" aria-hidden="true"></span></a>
      </div>
  <div id="left-column">
    <div class="sphinxsidebar">
<div class="sidebar-block">
  <div class="sidebar-wrapper">
    <h2>Contents</h2>
    <div class="sidebar-localtoc">
      <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Example Workflow</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#step-1-downloading-datasets">Step 1: Downloading datasets</a></li>
<li class="toctree-l2"><a class="reference internal" href="#step-2-copy-template">Step 2: Copy template</a></li>
<li class="toctree-l2"><a class="reference internal" href="#step-3-change-your-miniconda-path">Step 3 : Change your miniconda path</a></li>
<li class="toctree-l2"><a class="reference internal" href="#step-4-edit-snakemake-cluster-yaml">Step 4: Edit snakemake/cluster.yaml</a></li>
<li class="toctree-l2"><a class="reference internal" href="#step-5-submit-job">Step 5: Submit job</a></li>
<li class="toctree-l2"><a class="reference internal" href="#visualizing-results">Visualizing Results</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="cmd-manual.html">Manual</a></li>
<li class="toctree-l1"><a class="reference internal" href="api-usage.html">API Usage</a></li>
<li class="toctree-l1"><a class="reference internal" href="scores.html">Scores</a></li>
<li class="toctree-l1"><a class="reference internal" href="modules.html">riboraptor</a></li>
<li class="toctree-l1"><a class="reference internal" href="history.html">History</a></li>
<li class="toctree-l1"><a class="reference internal" href="contributing.html">Contributing</a></li>
<li class="toctree-l1"><a class="reference internal" href="authors.html">Credits</a></li>
</ul>

    </div>
  </div>
</div>
  <h4>Previous topic</h4>
  <p class="topless"><a href="installation.html"
                        title="previous chapter">Installation</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="cmd-manual.html"
                        title="next chapter">Manual</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/example_workflow.rst.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div class="sidebar-block">
  <div class="sidebar-wrapper">
    <div id="main-search">
      <form class="form-inline" action="search.html" method="GET" role="form">
        <div class="input-group">
          <input name="q" type="text" class="form-control" placeholder="Search...">
        </div>
        <input type="hidden" name="check_keywords" value="yes" />
        <input type="hidden" name="area" value="default" />
      </form>
    </div>
  </div>
</div>
      
    </div>
  </div>
        <div id="right-column">
          
          <div role="navigation" aria-label="breadcrumbs navigation">
            <ol class="breadcrumb">
              <li><a href="index.html">Docs</a></li>
              
              <li>Example Workflow</li>
            </ol>
          </div>
          
          <div class="document clearer body">
            
  <div class="section" id="example-workflow">
<h1>Example Workflow<a class="headerlink" href="#example-workflow" title="Permalink to this headline">¶</a></h1>
<p>We will be working with the first published Ribo-seq dataset <a class="reference external" href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE13750">GSE13750</a> from Ingolia et al. (2009) which has samples for <a class="reference external" href="https://www.ncbi.nlm.nih.gov/Traces/study/?acc=SRP000637">both</a> mRNA-seq and Ribo-seq from <strong>Yeast</strong> grown in starved and nutrient rich media.</p>
<p>At this point, we assume you have already completed all
the steps under “Installing dependencies” section of the README.</p>
<div class="section" id="step-1-downloading-datasets">
<h2>Step 1: Downloading datasets<a class="headerlink" href="#step-1-downloading-datasets" title="Permalink to this headline">¶</a></h2>
<p>We will download all SRA files corresponding to GSE13750.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="nb">cd</span> riboraptor
download_sra_data --sradb<span class="o">=</span>../riboraptor-data/SRAmetadb.sqlite <span class="se">\</span>
--geodb<span class="o">=</span>../riboraptor-data/GEOmetadb.sqlite GSE13750
</pre></div>
</div>
<p>GEO IDs are automatatiicaly converted to corresponding SRP IDs.
GSE13750 corresponds to SRP000637.</p>
<p>There are 6 experiments in total (SRX003184-SRX003191), but we will be working with only two: <cite>SRX003187`and `SRX003191</cite>
one of which is mRNA-seq while other is Ribo-seq. (We will figure out which is which later.)
You can delete are SRX directories except the above two. We will be using <cite>sacCerR64</cite> as our reference.</p>
<p>We will now use Snakemake to run all the downstream steps. Here is what the overall workflow looks like:</p>
<div class="figure">
<a class="reference internal image-reference" href="_images/dag.svg"><img alt="_images/dag.svg" src="_images/dag.svg" /></a>
</div>
</div>
<div class="section" id="step-2-copy-template">
<h2>Step 2: Copy template<a class="headerlink" href="#step-2-copy-template" title="Permalink to this headline">¶</a></h2>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="nb">cd</span> snakemake
cp configs/SRP000637.py.sample configs/SRP000637.py
</pre></div>
</div>
<p>Edit the paths inside <cite>SRP000637.py</cite> to point to your RAW data, GTF and BED files. BED files for a lot of assemblies are
inbuilt into riboraptor. In most cases, those should suffice. However, you stil need to provide paths to fasta, chromosome sizes and the GTF. The BED files are created from a particular version of the GTF, so if you are using your own GTF, you should ideally be using your own BED files too.</p>
<p>An example of a config would be:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">## Path to SRP directory</span>
<span class="n">RAWDATA_DIR</span> <span class="o">=</span> <span class="s1">&#39;/staging/as/skchoudh/SRA_datasets/SRP000637&#39;</span>

<span class="c1">## Output directory (will be created if does not exist)</span>
<span class="n">OUT_DIR</span> <span class="o">=</span> <span class="s1">&#39;/staging/as/skchoudh/riboraptor-analysis/SRP000637&#39;</span>

<span class="c1">## Genome fasta location</span>
<span class="n">GENOME_FASTA</span> <span class="o">=</span> <span class="s1">&#39;/home/cmb-06/as/skchoudh/genomes/sacCerR64/fasta/Saccharomyces_cerevisiae.R64-1-1.dna.toplevel.fa&#39;</span>

<span class="c1">## Chromosome sizes location</span>
<span class="n">CHROM_SIZES</span> <span class="o">=</span> <span class="s1">&#39;/home/cmb-06/as/skchoudh/genomes/sacCerR64/fasta/Saccharomyces_cerevisiae.R64-1-1.dna.toplevel.sizes&#39;</span>

<span class="c1">## Path to STAR index (will be generated if does not exist)</span>
<span class="n">STAR_INDEX</span> <span class="o">=</span> <span class="s1">&#39;/home/cmb-06/as/skchoudh/genomes/sacCerR64/star_annotated&#39;</span>

<span class="c1">## GTF path</span>
<span class="n">GTF</span> <span class="o">=</span> <span class="s1">&#39;/home/cmb-06/as/skchoudh/genomes/sacCerR64/annotation/Saccharomyces_cerevisiae.R64-1-1.91.gtf&#39;</span>


<span class="c1">## Path to bed file containing Intron coordinates</span>
<span class="n">INTRON_BED</span> <span class="o">=</span> <span class="s1">&#39;/home/cmb-panasas2/skchoudh/github_projects/riboraptor/riboraptor/annotation/sacCerR64/intron.bed&#39;</span>

<span class="c1">## Path to bed file containing CDS coordinates</span>
<span class="n">CDS_BED</span> <span class="o">=</span> <span class="s1">&#39;/home/cmb-panasas2/skchoudh/github_projects/riboraptor/riboraptor/annotation/sacCerR64/cds.bed&#39;</span>

<span class="c1">## Path to bed file containing 5&#39;UTR coordinates</span>
<span class="n">UTR5_BED</span> <span class="o">=</span> <span class="s1">&#39;/home/cmb-panasas2/skchoudh/github_projects/riboraptor/riboraptor/annotation/sacCerR64/utr5.bed&#39;</span>

<span class="c1">## Path to bed file containing 3&#39;UTR coordinates</span>
<span class="n">UTR3_BED</span> <span class="o">=</span> <span class="s1">&#39;/home/cmb-panasas2/skchoudh/github_projects/riboraptor/riboraptor/annotation/sacCerR64/utr3.bed&#39;</span>
</pre></div>
</div>
</div>
<div class="section" id="step-3-change-your-miniconda-path">
<h2>Step 3 : Change your miniconda path<a class="headerlink" href="#step-3-change-your-miniconda-path" title="Permalink to this headline">¶</a></h2>
<p>Edit <a class="reference external" href="https://github.com/saketkc/riboraptor/blob/47c8a50753c2bcc96b57d43b525a47bb8fde2d04/snakemake/jobscript.sh#L4">Line4 snakemake/jobscript.sh</a> pointing to your conda root directory.</p>
<blockquote>
<div>An example path would be:</div></blockquote>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="nb">export</span> <span class="nv">PATH</span><span class="o">=</span><span class="s2">&quot;/home/cmb-panasas2/wenzhenl/miniconda3/bin:</span><span class="nv">$PATH</span><span class="s2">&quot;</span>
</pre></div>
</div>
</div>
<div class="section" id="step-4-edit-snakemake-cluster-yaml">
<h2>Step 4: Edit snakemake/cluster.yaml<a class="headerlink" href="#step-4-edit-snakemake-cluster-yaml" title="Permalink to this headline">¶</a></h2>
<p>Edit <a class="reference external" href="https://github.com/saketkc/riboraptor/blob/47c8a50753c2bcc96b57d43b525a47bb8fde2d04/snakemake/cluster.yaml#L6">Line6 snakemake/cluster.yaml</a> and <a class="reference external" href="https://github.com/saketkc/riboraptor/blob/47c8a50753c2bcc96b57d43b525a47bb8fde2d04/snakemake/cluster.yaml#L7">Line7 snakemake/cluster.yaml</a> to point
to your log directory error log file.</p>
<blockquote>
<div>An example path would be:</div></blockquote>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="l l-Scalar l-Scalar-Plain">logout</span><span class="p p-Indicator">:</span> <span class="s">&#39;/home/cmb-06/as/skchoudh/logs/{rule}.{wildcards}.out&#39;</span>
<span class="l l-Scalar l-Scalar-Plain">logerror</span><span class="p p-Indicator">:</span> <span class="s">&#39;/home/cmb-06/as/skchoudh/logs/{rule}.{wildcards}.err&#39;</span>
</pre></div>
</div>
<p>You would want to just edit the directory path leading to <cite>/home/cmb-06/as/skchoudh/logs/</cite> and leave the rest as it is.</p>
</div>
<div class="section" id="step-5-submit-job">
<h2>Step 5: Submit job<a class="headerlink" href="#step-5-submit-job" title="Permalink to this headline">¶</a></h2>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>bash submitall.sh SRP000637
</pre></div>
</div>
<p>The <cite>submitall.sh</cite> looks for a file named <cite>SRP000637.py</cite> in the configs directory, so make sure <cite>SRP000637.py</cite> exists inside
<cite>configs/</cite> directory.</p>
</div>
<div class="section" id="visualizing-results">
<h2>Visualizing Results<a class="headerlink" href="#visualizing-results" title="Permalink to this headline">¶</a></h2>
<p>When the entire pipeline as run, it will create an html file <cite>riboraptor_report.html</cite> as output. You can copy it locally to visualize metagene profiles and read length distribution for all samples.</p>
</div>
</div>


          </div>
            
  <div class="footer-relations">
    
      <div class="pull-left">
        <a class="btn btn-default" href="installation.html" title="previous chapter (use the left arrow)">Installation</a>
      </div>
    
      <div class="pull-right">
        <a class="btn btn-default" href="cmd-manual.html" title="next chapter (use the right arrow)">Manual</a>
      </div>
    </div>
    <div class="clearer"></div>
  
        </div>
        <div class="clearfix"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="cmd-manual.html" title="Manual"
             >next</a> |</li>
        <li class="right" >
          <a href="installation.html" title="Installation"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">riboraptor 0.3.0 documentation</a> &#187;</li> 
      </ul>
    </div>
<script type="text/javascript">
  $("#mobile-toggle a").click(function () {
    $("#left-column").toggle();
  });
</script>
<script type="text/javascript" src="_static/js/bootstrap.js"></script>
  <div class="footer">
    &copy; Copyright 2018, Saket Choudhary. Created using <a href="http://sphinx.pocoo.org/">Sphinx</a>.
  </div>
  </body>
</html>