{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Populating the interactive namespace from numpy and matplotlib\n"
     ]
    }
   ],
   "source": [
    "%pylab inline\n",
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "import pandas as pd\n",
    "import os\n",
    "import fnmatch\n",
    "import errno    \n",
    "import glob\n",
    "from riboraptor.sradb import SRAdb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "db = SRAdb('/staging/as/skchoudh/SRAmetadb.sqlite')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# List tables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['metaInfo',\n",
       " 'submission',\n",
       " 'study',\n",
       " 'sample',\n",
       " 'experiment',\n",
       " 'run',\n",
       " 'sra',\n",
       " 'sra_ft',\n",
       " 'sra_ft_content',\n",
       " 'sra_ft_segments',\n",
       " 'sra_ft_segdir',\n",
       " 'col_desc',\n",
       " 'fastq']"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sra_tables = db.list_tables()\n",
    "sra_tables"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# List fields in a table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['study_ID',\n",
       " 'study_alias',\n",
       " 'study_accession',\n",
       " 'study_title',\n",
       " 'study_type',\n",
       " 'study_abstract',\n",
       " 'broker_name',\n",
       " 'center_name',\n",
       " 'center_project_name',\n",
       " 'study_description',\n",
       " 'related_studies',\n",
       " 'primary_study',\n",
       " 'sra_link',\n",
       " 'study_url_link',\n",
       " 'xref_link',\n",
       " 'study_entrez_link',\n",
       " 'ddbj_link',\n",
       " 'ena_link',\n",
       " 'study_attribute',\n",
       " 'submission_accession',\n",
       " 'sradb_updated']"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "db.list_fields('study')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Describe columns in a table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cid\tname\ttype\tnotnull\tdflt_value\tpk\n",
      "\n",
      "0\tstudy_ID\tREAL\t0\tNone\t0\n",
      "1\tstudy_alias\tTEXT\t0\tNone\t0\n",
      "2\tstudy_accession\tTEXT\t0\tNone\t0\n",
      "3\tstudy_title\tTEXT\t0\tNone\t0\n",
      "4\tstudy_type\tTEXT\t0\tNone\t0\n",
      "5\tstudy_abstract\tTEXT\t0\tNone\t0\n",
      "6\tbroker_name\tTEXT\t0\tNone\t0\n",
      "7\tcenter_name\tTEXT\t0\tNone\t0\n",
      "8\tcenter_project_name\tTEXT\t0\tNone\t0\n",
      "9\tstudy_description\tTEXT\t0\tNone\t0\n",
      "10\trelated_studies\tTEXT\t0\tNone\t0\n",
      "11\tprimary_study\tTEXT\t0\tNone\t0\n",
      "12\tsra_link\tTEXT\t0\tNone\t0\n",
      "13\tstudy_url_link\tTEXT\t0\tNone\t0\n",
      "14\txref_link\tTEXT\t0\tNone\t0\n",
      "15\tstudy_entrez_link\tTEXT\t0\tNone\t0\n",
      "16\tddbj_link\tTEXT\t0\tNone\t0\n",
      "17\tena_link\tTEXT\t0\tNone\t0\n",
      "18\tstudy_attribute\tTEXT\t0\tNone\t0\n",
      "19\tsubmission_accession\tTEXT\t0\tNone\t0\n",
      "20\tsradb_updated\tTEXT\t0\tNone\t0\n"
     ]
    }
   ],
   "source": [
    "db.desc_table('study')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Query"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>broker_name</th>\n",
       "      <th>center_name</th>\n",
       "      <th>center_project_name</th>\n",
       "      <th>ddbj_link</th>\n",
       "      <th>ena_link</th>\n",
       "      <th>primary_study</th>\n",
       "      <th>related_studies</th>\n",
       "      <th>sra_link</th>\n",
       "      <th>sradb_updated</th>\n",
       "      <th>study_ID</th>\n",
       "      <th>...</th>\n",
       "      <th>study_accession</th>\n",
       "      <th>study_alias</th>\n",
       "      <th>study_attribute</th>\n",
       "      <th>study_description</th>\n",
       "      <th>study_entrez_link</th>\n",
       "      <th>study_title</th>\n",
       "      <th>study_type</th>\n",
       "      <th>study_url_link</th>\n",
       "      <th>submission_accession</th>\n",
       "      <th>xref_link</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>None</td>\n",
       "      <td>KEIO</td>\n",
       "      <td>Bacillus subtilis subsp. natto BEST195</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>2016-11-20 16:42:15</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>DRP000001</td>\n",
       "      <td>DRP000001</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Bacillus subtilis subsp. natto BEST195 genome ...</td>\n",
       "      <td>Whole Genome Sequencing</td>\n",
       "      <td>None</td>\n",
       "      <td>DRA000001</td>\n",
       "      <td>pubmed: 20398357 || pubmed: 25329997</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>None</td>\n",
       "      <td>KEIO</td>\n",
       "      <td>Bacillus subtilis subsp. subtilis str. 168</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>2017-09-21 20:23:35</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>DRP000002</td>\n",
       "      <td>DRP000002</td>\n",
       "      <td>None</td>\n",
       "      <td>Whole genome resequencing of B. subtilis subti...</td>\n",
       "      <td>None</td>\n",
       "      <td>Model organism for prokaryotic cell differenti...</td>\n",
       "      <td>Whole Genome Sequencing</td>\n",
       "      <td>None</td>\n",
       "      <td>DRA000002</td>\n",
       "      <td>pubmed: 20398357</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>None</td>\n",
       "      <td>UT-MGS</td>\n",
       "      <td>Integrative Transcriptome Analysis</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>2016-11-20 16:42:16</td>\n",
       "      <td>3.0</td>\n",
       "      <td>...</td>\n",
       "      <td>DRP000003</td>\n",
       "      <td>DRP000003</td>\n",
       "      <td>None</td>\n",
       "      <td>Although recent studies have revealed that the...</td>\n",
       "      <td>None</td>\n",
       "      <td>Comprehensive identification and characterizat...</td>\n",
       "      <td>Transcriptome Analysis</td>\n",
       "      <td>DBTSS: http://dbtss.hgc.jp/</td>\n",
       "      <td>DRA000003</td>\n",
       "      <td>pubmed: 20400770</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows Ã— 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  broker_name center_name                         center_project_name  \\\n",
       "0        None        KEIO      Bacillus subtilis subsp. natto BEST195   \n",
       "1        None        KEIO  Bacillus subtilis subsp. subtilis str. 168   \n",
       "2        None      UT-MGS          Integrative Transcriptome Analysis   \n",
       "\n",
       "  ddbj_link ena_link primary_study related_studies sra_link  \\\n",
       "0      None     None          None            None     None   \n",
       "1      None     None          None            None     None   \n",
       "2      None     None          None            None     None   \n",
       "\n",
       "         sradb_updated  study_ID                  ...                   \\\n",
       "0  2016-11-20 16:42:15       1.0                  ...                    \n",
       "1  2017-09-21 20:23:35       2.0                  ...                    \n",
       "2  2016-11-20 16:42:16       3.0                  ...                    \n",
       "\n",
       "  study_accession study_alias study_attribute  \\\n",
       "0       DRP000001   DRP000001            None   \n",
       "1       DRP000002   DRP000002            None   \n",
       "2       DRP000003   DRP000003            None   \n",
       "\n",
       "                                   study_description study_entrez_link  \\\n",
       "0                                               None              None   \n",
       "1  Whole genome resequencing of B. subtilis subti...              None   \n",
       "2  Although recent studies have revealed that the...              None   \n",
       "\n",
       "                                         study_title               study_type  \\\n",
       "0  Bacillus subtilis subsp. natto BEST195 genome ...  Whole Genome Sequencing   \n",
       "1  Model organism for prokaryotic cell differenti...  Whole Genome Sequencing   \n",
       "2  Comprehensive identification and characterizat...   Transcriptome Analysis   \n",
       "\n",
       "                study_url_link submission_accession  \\\n",
       "0                         None            DRA000001   \n",
       "1                         None            DRA000002   \n",
       "2  DBTSS: http://dbtss.hgc.jp/            DRA000003   \n",
       "\n",
       "                              xref_link  \n",
       "0  pubmed: 20398357 || pubmed: 25329997  \n",
       "1                      pubmed: 20398357  \n",
       "2                      pubmed: 20400770  \n",
       "\n",
       "[3 rows x 21 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "db.get_query('SELECT * FROM study LIMIT 3')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Number of rows in each table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>metaInfo</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample</th>\n",
       "      <td>4199180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>submission</th>\n",
       "      <td>961295</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sra_ft_content</th>\n",
       "      <td>5242098</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>run</th>\n",
       "      <td>5339443</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>study</th>\n",
       "      <td>155648</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>fastq</th>\n",
       "      <td>5252155</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sra</th>\n",
       "      <td>5242098</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sra_ft_segdir</th>\n",
       "      <td>984</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sra_ft_segments</th>\n",
       "      <td>2964125</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>col_desc</th>\n",
       "      <td>129</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sra_ft</th>\n",
       "      <td>5242098</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>experiment</th>\n",
       "      <td>4708115</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   count\n",
       "metaInfo               2\n",
       "sample           4199180\n",
       "submission        961295\n",
       "sra_ft_content   5242098\n",
       "run              5339443\n",
       "study             155648\n",
       "fastq            5252155\n",
       "sra              5242098\n",
       "sra_ft_segdir        984\n",
       "sra_ft_segments  2964125\n",
       "col_desc             129\n",
       "sra_ft           5242098\n",
       "experiment       4708115"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "db.get_table_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Which study types are the most prominent?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Number</th>\n",
       "      <th>StudyType</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>45967</td>\n",
       "      <td>Whole Genome Sequencing</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>38853</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>38512</td>\n",
       "      <td>Other</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>16149</td>\n",
       "      <td>Transcriptome Analysis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>14415</td>\n",
       "      <td>Metagenomics</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>763</td>\n",
       "      <td>Population Genomics</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>607</td>\n",
       "      <td>Epigenetics</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>230</td>\n",
       "      <td>Exome Sequencing</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>110</td>\n",
       "      <td>Cancer Genomics</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>31</td>\n",
       "      <td>Pooled Clone Sequencing</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>9</td>\n",
       "      <td>Synthetic Genomics</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>1</td>\n",
       "      <td>Transcriptome Sequencing</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>1</td>\n",
       "      <td>Whole Genome Sequencing</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Number                 StudyType\n",
       "0    45967   Whole Genome Sequencing\n",
       "1    38853                      None\n",
       "2    38512                     Other\n",
       "3    16149    Transcriptome Analysis\n",
       "4    14415              Metagenomics\n",
       "5      763       Population Genomics\n",
       "6      607               Epigenetics\n",
       "7      230          Exome Sequencing\n",
       "8      110           Cancer Genomics\n",
       "9       31   Pooled Clone Sequencing\n",
       "10       9        Synthetic Genomics\n",
       "11       1  Transcriptome Sequencing\n",
       "12       1  Whole Genome Sequencing "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "db.get_query('SELECT study_type AS StudyType, count(*) AS Number FROM \"study\" GROUP BY study_type order by Number DESC')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Which Sequencing Instruments are most common?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Experiments</th>\n",
       "      <th>Instrument Model</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>8704</td>\n",
       "      <td>454 GS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>986</td>\n",
       "      <td>454 GS 20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>49413</td>\n",
       "      <td>454 GS FLX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>10</td>\n",
       "      <td>454 GS FLX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>148483</td>\n",
       "      <td>454 GS FLX Titanium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>16702</td>\n",
       "      <td>454 GS FLX+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>23956</td>\n",
       "      <td>454 GS Junior</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>1</td>\n",
       "      <td>AB 310 Genetic Analyzer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>111</td>\n",
       "      <td>AB 3130 Genetic Analyzer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>55</td>\n",
       "      <td>AB 3130xL Genetic Analyzer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>71</td>\n",
       "      <td>AB 3500 Genetic Analyzer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>14</td>\n",
       "      <td>AB 3500xL Genetic Analyzer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>260</td>\n",
       "      <td>AB 3730 Genetic Analyzer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>1168</td>\n",
       "      <td>AB 3730xL Genetic Analyzer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>2111</td>\n",
       "      <td>AB 5500 Genetic Analyzer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>5138</td>\n",
       "      <td>AB 5500xl Genetic Analyzer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>203</td>\n",
       "      <td>AB 5500xl-W Genetic Analysis System</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>323</td>\n",
       "      <td>AB SOLiD 3 Plus System</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>11025</td>\n",
       "      <td>AB SOLiD 4 System</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>153</td>\n",
       "      <td>AB SOLiD 4hq System</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>3</td>\n",
       "      <td>AB SOLiD PI System</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>501</td>\n",
       "      <td>AB SOLiD System</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>483</td>\n",
       "      <td>AB SOLiD System 2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>2581</td>\n",
       "      <td>AB SOLiD System 3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>319</td>\n",
       "      <td>BGISEQ-500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>4276</td>\n",
       "      <td>Complete Genomics</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>3937</td>\n",
       "      <td>Helicos HeliScope</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>18</td>\n",
       "      <td>HiSeq X Five</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>53240</td>\n",
       "      <td>HiSeq X Ten</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>19075</td>\n",
       "      <td>Illumina Genome Analyzer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>112444</td>\n",
       "      <td>Illumina Genome Analyzer II</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>63587</td>\n",
       "      <td>Illumina Genome Analyzer IIx</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>2</td>\n",
       "      <td>Illumina Genome Analyzer IIx</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>3675</td>\n",
       "      <td>Illumina HiScanSQ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>14584</td>\n",
       "      <td>Illumina HiSeq 1000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>9759</td>\n",
       "      <td>Illumina HiSeq 1500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1352215</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>572648</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>6082</td>\n",
       "      <td>Illumina HiSeq 3000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>29911</td>\n",
       "      <td>Illumina HiSeq 4000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>634091</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>1</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>361</td>\n",
       "      <td>Illumina MiniSeq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>18</td>\n",
       "      <td>Illumina NextSeq 500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>41</td>\n",
       "      <td>Ion S5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>31574</td>\n",
       "      <td>Ion Torrent PGM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>8420</td>\n",
       "      <td>Ion Torrent Proton</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>857</td>\n",
       "      <td>MinION</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>89496</td>\n",
       "      <td>NextSeq 500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>3259</td>\n",
       "      <td>NextSeq 550</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>12056</td>\n",
       "      <td>PacBio RS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>14435</td>\n",
       "      <td>PacBio RS II</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>53</td>\n",
       "      <td>Sequel</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>32933</td>\n",
       "      <td>unspecified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1362293</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Experiments                     Instrument Model\n",
       "21         8704                               454 GS\n",
       "32          986                            454 GS 20\n",
       "9         49413                           454 GS FLX\n",
       "50           10                          454 GS FLX \n",
       "4        148483                  454 GS FLX Titanium\n",
       "15        16702                          454 GS FLX+\n",
       "13        23956                        454 GS Junior\n",
       "53            1              AB 310 Genetic Analyzer\n",
       "42          111             AB 3130 Genetic Analyzer\n",
       "44           55           AB 3130xL Genetic Analyzer\n",
       "43           71             AB 3500 Genetic Analyzer\n",
       "49           14           AB 3500xL Genetic Analyzer\n",
       "39          260             AB 3730 Genetic Analyzer\n",
       "31         1168           AB 3730xL Genetic Analyzer\n",
       "30         2111             AB 5500 Genetic Analyzer\n",
       "24         5138           AB 5500xl Genetic Analyzer\n",
       "40          203  AB 5500xl-W Genetic Analysis System\n",
       "37          323               AB SOLiD 3 Plus System\n",
       "19        11025                    AB SOLiD 4 System\n",
       "41          153                  AB SOLiD 4hq System\n",
       "51            3                   AB SOLiD PI System\n",
       "34          501                      AB SOLiD System\n",
       "35          483                  AB SOLiD System 2.0\n",
       "29         2581                  AB SOLiD System 3.0\n",
       "38          319                           BGISEQ-500\n",
       "25         4276                    Complete Genomics\n",
       "26         3937                    Helicos HeliScope\n",
       "47           18                         HiSeq X Five\n",
       "8         53240                          HiSeq X Ten\n",
       "14        19075             Illumina Genome Analyzer\n",
       "5        112444          Illumina Genome Analyzer II\n",
       "7         63587         Illumina Genome Analyzer IIx\n",
       "52            2        Illumina Genome Analyzer IIx \n",
       "27         3675                    Illumina HiScanSQ\n",
       "16        14584                  Illumina HiSeq 1000\n",
       "20         9759                  Illumina HiSeq 1500\n",
       "1       1352215                  Illumina HiSeq 2000\n",
       "3        572648                  Illumina HiSeq 2500\n",
       "23         6082                  Illumina HiSeq 3000\n",
       "12        29911                  Illumina HiSeq 4000\n",
       "2        634091                       Illumina MiSeq\n",
       "54            1                      Illumina MiSeq \n",
       "36          361                     Illumina MiniSeq\n",
       "48           18                 Illumina NextSeq 500\n",
       "46           41                               Ion S5\n",
       "11        31574                      Ion Torrent PGM\n",
       "22         8420                   Ion Torrent Proton\n",
       "33          857                               MinION\n",
       "6         89496                          NextSeq 500\n",
       "28         3259                          NextSeq 550\n",
       "18        12056                            PacBio RS\n",
       "17        14435                         PacBio RS II\n",
       "45           53                               Sequel\n",
       "10        32933                          unspecified\n",
       "0       1362293                                 None"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "db.get_query('SELECT instrument_model AS \"Instrument Model\", count( * ) AS Experiments FROM \"experiment\" GROUP BY instrument_model order by Experiments DESC').sort_values(by=['Instrument Model'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Which assays are most common?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Library Strategy</th>\n",
       "      <th>Runs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>None</td>\n",
       "      <td>1362293</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>WGS</td>\n",
       "      <td>1092229</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>690441</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>627617</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>OTHER</td>\n",
       "      <td>314145</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>WXS</td>\n",
       "      <td>237358</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>CLONE</td>\n",
       "      <td>89555</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ChIP-Seq</td>\n",
       "      <td>82832</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>POOLCLONE</td>\n",
       "      <td>53738</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Bisulfite-Seq</td>\n",
       "      <td>32335</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>SELEX</td>\n",
       "      <td>25875</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>miRNA-Seq</td>\n",
       "      <td>22405</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>WGA</td>\n",
       "      <td>21078</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>RAD-Seq</td>\n",
       "      <td>10290</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Targeted-Capture</td>\n",
       "      <td>8427</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ATAC-seq</td>\n",
       "      <td>6060</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>ncRNA-Seq</td>\n",
       "      <td>5360</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>EST</td>\n",
       "      <td>3601</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>RIP-Seq</td>\n",
       "      <td>3264</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>DNase-Hypersensitivity</td>\n",
       "      <td>2868</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>MeDIP-Seq</td>\n",
       "      <td>2442</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>MNase-Seq</td>\n",
       "      <td>2328</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>MRE-Seq</td>\n",
       "      <td>2073</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>FL-cDNA</td>\n",
       "      <td>2010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Tn-Seq</td>\n",
       "      <td>1886</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>WCS</td>\n",
       "      <td>1809</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>MBD-Seq</td>\n",
       "      <td>1690</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>CLONEEND</td>\n",
       "      <td>523</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>FAIRE-seq</td>\n",
       "      <td>477</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>Hi-C</td>\n",
       "      <td>323</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>CTS</td>\n",
       "      <td>282</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>other</td>\n",
       "      <td>214</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>Synthetic-Long-Read</td>\n",
       "      <td>195</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>FINISHING</td>\n",
       "      <td>40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>ChIA-PET</td>\n",
       "      <td>30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>VALIDATION</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          Library Strategy     Runs\n",
       "0                     None  1362293\n",
       "1                      WGS  1092229\n",
       "2                 AMPLICON   690441\n",
       "3                  RNA-Seq   627617\n",
       "4                    OTHER   314145\n",
       "5                      WXS   237358\n",
       "6                    CLONE    89555\n",
       "7                 ChIP-Seq    82832\n",
       "8                POOLCLONE    53738\n",
       "9            Bisulfite-Seq    32335\n",
       "10                   SELEX    25875\n",
       "11               miRNA-Seq    22405\n",
       "12                     WGA    21078\n",
       "13                 RAD-Seq    10290\n",
       "14        Targeted-Capture     8427\n",
       "15                ATAC-seq     6060\n",
       "16               ncRNA-Seq     5360\n",
       "17                     EST     3601\n",
       "18                 RIP-Seq     3264\n",
       "19  DNase-Hypersensitivity     2868\n",
       "20               MeDIP-Seq     2442\n",
       "21               MNase-Seq     2328\n",
       "22                 MRE-Seq     2073\n",
       "23                 FL-cDNA     2010\n",
       "24                  Tn-Seq     1886\n",
       "25                     WCS     1809\n",
       "26                 MBD-Seq     1690\n",
       "27                CLONEEND      523\n",
       "28               FAIRE-seq      477\n",
       "29                    Hi-C      323\n",
       "30                     CTS      282\n",
       "31                   other      214\n",
       "32     Synthetic-Long-Read      195\n",
       "33               FINISHING       40\n",
       "34                ChIA-PET       30\n",
       "35              VALIDATION       22"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "db.get_query('SELECT library_strategy AS \"Library Strategy\", count( * ) AS Runs FROM \"experiment\" GROUP BY library_strategy order by Runs DESC')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (riboraptor_clean)",
   "language": "python",
   "name": "riboraptor_clean"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
