
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="X-UA-Compatible" content="IE=Edge" />
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
  
  <!-- Licensed under the Apache 2.0 License -->
  <link rel="stylesheet" type="text/css" href="../../_static/fonts/open-sans/stylesheet.css" />
  <!-- Licensed under the SIL Open Font License -->
  <link rel="stylesheet" type="text/css" href="../../_static/fonts/source-serif-pro/source-serif-pro.css" />
  <link rel="stylesheet" type="text/css" href="../../_static/css/bootstrap.min.css" />
  <link rel="stylesheet" type="text/css" href="../../_static/css/bootstrap-theme.min.css" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
    <title>riboraptor.hdf_parser &#8212; riboraptor 0.3.0 documentation</title>
    <link rel="stylesheet" href="../../_static/guzzle.css" type="text/css" />
    <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../../_static/graphviz.css" />
    <script type="text/javascript" id="documentation_options" data-url_root="../../" src="../../_static/documentation_options.js"></script>
    <script type="text/javascript" src="../../_static/jquery.js"></script>
    <script type="text/javascript" src="../../_static/underscore.js"></script>
    <script type="text/javascript" src="../../_static/doctools.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
  
   

  </head><body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="../../index.html">riboraptor 0.3.0 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="../index.html" accesskey="U">Module code</a> &#187;</li> 
      </ul>
    </div>
    <div class="container-wrapper">

      <div id="mobile-toggle">
        <a href="#"><span class="glyphicon glyphicon-align-justify" aria-hidden="true"></span></a>
      </div>
  <div id="left-column">
    <div class="sphinxsidebar">
<div class="sidebar-block">
  <div class="sidebar-wrapper">
    <h2>Contents</h2>
    <div class="sidebar-localtoc">
      <ul>
<li class="toctree-l1"><a class="reference internal" href="../../installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../example_workflow.html">Example Workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../cmd-manual.html">Manual</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../api-usage.html">API Usage</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../scores.html">Scores</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../modules.html">riboraptor</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../history.html">History</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../contributing.html">Contributing</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../authors.html">Credits</a></li>
</ul>

    </div>
  </div>
</div>
<div class="sidebar-block">
  <div class="sidebar-wrapper">
    <div id="main-search">
      <form class="form-inline" action="../../search.html" method="GET" role="form">
        <div class="input-group">
          <input name="q" type="text" class="form-control" placeholder="Search...">
        </div>
        <input type="hidden" name="check_keywords" value="yes" />
        <input type="hidden" name="area" value="default" />
      </form>
    </div>
  </div>
</div>
      
    </div>
  </div>
        <div id="right-column">
          
          <div role="navigation" aria-label="breadcrumbs navigation">
            <ol class="breadcrumb">
              <li><a href="../../index.html">Docs</a></li>
              
                <li><a href="../index.html">Module code</a></li>
              
              <li>riboraptor.hdf_parser</li>
            </ol>
          </div>
          
          <div class="document clearer body">
            
  <h1>Source code for riboraptor.hdf_parser</h1><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">__future__</span> <span class="k">import</span> <span class="n">absolute_import</span>
<span class="kn">from</span> <span class="nn">__future__</span> <span class="k">import</span> <span class="n">division</span>
<span class="kn">from</span> <span class="nn">__future__</span> <span class="k">import</span> <span class="n">print_function</span>
<span class="kn">from</span> <span class="nn">__future__</span> <span class="k">import</span> <span class="n">unicode_literals</span>

<span class="kn">from</span> <span class="nn">collections</span> <span class="k">import</span> <span class="n">Counter</span>
<span class="kn">from</span> <span class="nn">collections</span> <span class="k">import</span> <span class="n">defaultdict</span>
<span class="kn">from</span> <span class="nn">collections</span> <span class="k">import</span> <span class="n">OrderedDict</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">subprocess</span>

<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">pybedtools</span>
<span class="kn">import</span> <span class="nn">h5py</span>
<span class="kn">import</span> <span class="nn">six</span>
<span class="kn">import</span> <span class="nn">pyBigWig</span>
<span class="kn">from</span> <span class="nn">tqdm</span> <span class="k">import</span> <span class="n">tqdm</span>
<span class="kn">import</span> <span class="nn">warnings</span>
<span class="kn">import</span> <span class="nn">sys</span>
<span class="kn">from</span> <span class="nn">joblib</span> <span class="k">import</span> <span class="n">delayed</span>

<span class="kn">from</span> <span class="nn">.genome</span> <span class="k">import</span> <span class="n">_get_bed</span>
<span class="kn">from</span> <span class="nn">.genome</span> <span class="k">import</span> <span class="n">__GENOMES_DB__</span>
<span class="kn">from</span> <span class="nn">.interval</span> <span class="k">import</span> <span class="n">Interval</span>

<span class="kn">from</span> <span class="nn">.helpers</span> <span class="k">import</span> <span class="n">complementary_strand</span>
<span class="kn">from</span> <span class="nn">.helpers</span> <span class="k">import</span> <span class="n">mkdir_p</span>
<span class="kn">from</span> <span class="nn">.helpers</span> <span class="k">import</span> <span class="n">merge_intervals</span>
<span class="kn">from</span> <span class="nn">.helpers</span> <span class="k">import</span> <span class="n">path_leaf</span>
<span class="kn">from</span> <span class="nn">.parallel</span> <span class="k">import</span> <span class="n">ParallelExecutor</span>
<span class="kn">from</span> <span class="nn">.count</span> <span class="k">import</span> <span class="n">_multiprocess_gene_coverage</span>
<span class="kn">from</span> <span class="nn">.helpers</span> <span class="k">import</span> <span class="n">scale_bigwig</span>
<span class="kn">import</span> <span class="nn">tempfile</span>
<span class="n">TMP_DIR_ROOT</span> <span class="o">=</span> <span class="s1">&#39;/tmp&#39;</span>

<span class="c1">#from multiprocessing import Pool</span>


<span class="k">def</span> <span class="nf">_scale_value_to_rpm</span><span class="p">(</span><span class="n">value</span><span class="p">,</span> <span class="n">total</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Scale value to reads per million&quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">value</span> <span class="o">*</span> <span class="mi">1</span> <span class="o">/</span> <span class="p">(</span><span class="n">total</span> <span class="o">/</span> <span class="mf">1e6</span><span class="p">)</span>


<span class="k">def</span> <span class="nf">_create_bigwig_from_bed</span><span class="p">(</span><span class="n">bed</span><span class="p">,</span> <span class="n">chrom_lengths</span><span class="p">,</span> <span class="n">outfile</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Create bigwig from a bed</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    bed: DataFrame</span>
<span class="sd">        Dataframe with standard six columns with appropriate names</span>
<span class="sd">    chrom_lengths: list of tuples</span>
<span class="sd">                   Sorted list of chromosome names and sizes</span>
<span class="sd">    outfile: string</span>
<span class="sd">             Path to store output file</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">bed</span><span class="p">,</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">):</span>
        <span class="n">bed</span> <span class="o">=</span> <span class="n">pybedtools</span><span class="o">.</span><span class="n">BedTool</span><span class="p">(</span><span class="n">bed</span><span class="p">)</span><span class="o">.</span><span class="n">as_dataframe</span><span class="p">()</span>

    <span class="n">bw</span> <span class="o">=</span> <span class="n">pyBigWig</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="n">outfile</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span>
    <span class="n">bw</span><span class="o">.</span><span class="n">addHeader</span><span class="p">(</span><span class="n">chrom_lengths</span><span class="p">,</span> <span class="n">maxZooms</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
    <span class="n">bw</span><span class="o">.</span><span class="n">addEntries</span><span class="p">(</span>
        <span class="n">bed</span><span class="p">[</span><span class="s1">&#39;chrom&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">(),</span>
        <span class="n">bed</span><span class="p">[</span><span class="s1">&#39;start&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">(),</span>
        <span class="n">ends</span><span class="o">=</span><span class="n">bed</span><span class="p">[</span><span class="s1">&#39;end&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">(),</span>
        <span class="n">values</span><span class="o">=</span><span class="n">bed</span><span class="p">[</span><span class="s1">&#39;score&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">())</span>
    <span class="n">bw</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>


<div class="viewcode-block" id="hdf_to_bigwig"><a class="viewcode-back" href="../../riboraptor.html#riboraptor.hdf_parser.hdf_to_bigwig">[docs]</a><span class="k">def</span> <span class="nf">hdf_to_bigwig</span><span class="p">(</span><span class="n">hdf</span><span class="p">,</span>
                  <span class="n">prefixdir</span><span class="p">,</span>
                  <span class="n">read_lengths_to_use</span><span class="o">=</span><span class="s1">&#39;all&#39;</span><span class="p">,</span>
                  <span class="n">output_normalized</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Create fragment and strand specific bigwigs from hdf</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    hdf: string</span>
<span class="sd">         Path to hdf file</span>
<span class="sd">    prefix: string</span>
<span class="sd">            Prefix to store output files</span>
<span class="sd">    read_lengths_to_use: list</span>
<span class="sd">                         Create bigwig only from these fragment lengths</span>
<span class="sd">    TODO: one fix for this would be to create separate nonoverlapping bigwigs</span>
<span class="sd">    and then merge them with bigWigCat (one more dependency)</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">hdf</span> <span class="o">=</span> <span class="n">h5py</span><span class="o">.</span><span class="n">File</span><span class="p">(</span><span class="n">hdf</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span>
    <span class="n">all_read_lengths</span> <span class="o">=</span> <span class="n">hdf</span><span class="p">[</span><span class="s1">&#39;read_lengths&#39;</span><span class="p">]</span>
    <span class="n">read_counts</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">hdf</span><span class="p">[</span><span class="s1">&#39;read_lengths_counts&#39;</span><span class="p">])</span>
    <span class="n">read_counts</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="n">read_counts</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="n">all_read_lengths</span><span class="p">)</span><span class="o">.</span><span class="n">sort_index</span><span class="p">()</span>
    <span class="n">chrom_names</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">map</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="nb">str</span><span class="p">(</span><span class="n">x</span><span class="p">),</span> <span class="n">hdf</span><span class="p">[</span><span class="s1">&#39;chrom_names&#39;</span><span class="p">]))</span>
    <span class="n">protocol</span> <span class="o">=</span> <span class="n">hdf</span><span class="o">.</span><span class="n">attrs</span><span class="p">[</span><span class="s1">&#39;protocol&#39;</span><span class="p">]</span>
    <span class="n">chrom_sizes</span> <span class="o">=</span> <span class="n">hdf</span><span class="p">[</span><span class="s1">&#39;chrom_sizes&#39;</span><span class="p">]</span>
    <span class="c1"># This is already sorted in the file</span>
    <span class="c1"># So no need to sort again (as required by bigwig)</span>
    <span class="n">chrom_lengths</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">chrom_names</span><span class="p">,</span> <span class="n">chrom_sizes</span><span class="p">))</span>
    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">read_lengths_to_use</span><span class="p">,</span> <span class="nb">int</span><span class="p">):</span>
        <span class="c1"># Make it a list</span>
        <span class="n">read_lengths_to_use</span> <span class="o">=</span> <span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="n">read_lengths_to_use</span><span class="p">)]</span>
    <span class="k">if</span> <span class="n">read_lengths_to_use</span> <span class="o">==</span> <span class="s1">&#39;all&#39;</span><span class="p">:</span>
        <span class="n">read_lengths</span> <span class="o">=</span> <span class="n">all_read_lengths</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">read_lengths</span> <span class="o">=</span> <span class="n">read_lengths_to_use</span>
    <span class="n">mkdir_p</span><span class="p">(</span><span class="n">prefixdir</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">read_length</span> <span class="ow">in</span> <span class="n">read_lengths</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">read_length</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">all_read_lengths</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">{}</span><span class="s1"> not in hdf&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">read_length</span><span class="p">))</span>
            <span class="k">continue</span>
        <span class="n">read_total</span> <span class="o">=</span> <span class="n">read_counts</span><span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="n">read_length</span><span class="p">)]</span>
        <span class="n">read_len_group</span> <span class="o">=</span> <span class="n">hdf</span><span class="p">[</span><span class="s1">&#39;fragments&#39;</span><span class="p">][</span><span class="n">read_length</span><span class="p">]</span>
        <span class="k">for</span> <span class="n">orientation</span> <span class="ow">in</span> <span class="n">hdf</span><span class="p">[</span><span class="s1">&#39;fragments&#39;</span><span class="p">][</span><span class="n">read_length</span><span class="p">]</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
            <span class="n">orientation_group</span> <span class="o">=</span> <span class="n">read_len_group</span><span class="p">[</span><span class="n">orientation</span><span class="p">]</span>
            <span class="n">dest_dir</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">abspath</span><span class="p">(</span>
                <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">prefixdir</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">read_length</span><span class="p">)))</span>
            <span class="n">mkdir_p</span><span class="p">(</span><span class="n">dest_dir</span><span class="p">)</span>
            <span class="n">bws</span> <span class="o">=</span> <span class="p">{}</span>
            <span class="n">bws</span><span class="p">[</span><span class="s1">&#39;pos_bw&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">dest_dir</span><span class="p">,</span> <span class="s1">&#39;</span><span class="si">{}</span><span class="s1">_</span><span class="si">{}</span><span class="s1">.bw&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                <span class="n">orientation</span><span class="p">,</span> <span class="s1">&#39;pos&#39;</span><span class="p">))</span>
            <span class="n">bws</span><span class="p">[</span><span class="s1">&#39;neg_bw&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">dest_dir</span><span class="p">,</span> <span class="s1">&#39;</span><span class="si">{}</span><span class="s1">_</span><span class="si">{}</span><span class="s1">.bw&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                <span class="n">orientation</span><span class="p">,</span> <span class="s1">&#39;neg&#39;</span><span class="p">))</span>
            <span class="c1"># This flle will store only the relevant</span>
            <span class="c1"># strand information</span>
            <span class="n">bws</span><span class="p">[</span><span class="s1">&#39;collapsed_bw&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
                <span class="n">dest_dir</span><span class="p">,</span> <span class="s1">&#39;</span><span class="si">{}</span><span class="s1">_</span><span class="si">{}</span><span class="s1">.bw&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">orientation</span><span class="p">,</span> <span class="s1">&#39;collapsed&#39;</span><span class="p">))</span>
            <span class="n">bws</span><span class="p">[</span><span class="s1">&#39;combined_bw&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
                <span class="n">dest_dir</span><span class="p">,</span> <span class="s1">&#39;</span><span class="si">{}</span><span class="s1">_</span><span class="si">{}</span><span class="s1">.bw&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">orientation</span><span class="p">,</span> <span class="s1">&#39;combined&#39;</span><span class="p">))</span>
            <span class="k">if</span> <span class="n">output_normalized</span><span class="p">:</span>
                <span class="n">bws</span><span class="p">[</span><span class="s1">&#39;pos_norm_bw&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
                    <span class="n">dest_dir</span><span class="p">,</span> <span class="s1">&#39;</span><span class="si">{}</span><span class="s1">_</span><span class="si">{}</span><span class="s1">.bw&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">orientation</span><span class="p">,</span> <span class="s1">&#39;pos_normalized&#39;</span><span class="p">))</span>
                <span class="n">bws</span><span class="p">[</span><span class="s1">&#39;neg_norm_bw&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
                    <span class="n">dest_dir</span><span class="p">,</span> <span class="s1">&#39;</span><span class="si">{}</span><span class="s1">_</span><span class="si">{}</span><span class="s1">.bw&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">orientation</span><span class="p">,</span> <span class="s1">&#39;neg_normalized&#39;</span><span class="p">))</span>
                <span class="n">bws</span><span class="p">[</span><span class="s1">&#39;combined_norm_bw&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
                    <span class="n">dest_dir</span><span class="p">,</span> <span class="s1">&#39;</span><span class="si">{}</span><span class="s1">_</span><span class="si">{}</span><span class="s1">.bw&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">orientation</span><span class="p">,</span>
                                                <span class="s1">&#39;combined_normalized&#39;</span><span class="p">))</span>

            <span class="k">for</span> <span class="n">key</span><span class="p">,</span> <span class="n">bw</span> <span class="ow">in</span> <span class="n">six</span><span class="o">.</span><span class="n">iteritems</span><span class="p">(</span><span class="n">bws</span><span class="p">):</span>
                <span class="n">bws</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">=</span> <span class="n">pyBigWig</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="n">bw</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span>
                <span class="n">bws</span><span class="p">[</span><span class="n">key</span><span class="p">]</span><span class="o">.</span><span class="n">addHeader</span><span class="p">(</span><span class="n">chrom_lengths</span><span class="p">,</span> <span class="n">maxZooms</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
            <span class="n">chrom_strand</span> <span class="o">=</span> <span class="nb">list</span><span class="p">()</span>
            <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">orientation_group</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
                <span class="n">chrom_strand</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">c</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;__&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">],</span> <span class="n">c</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;__&#39;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]))</span>

            <span class="n">chrom_strand</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">chrom_strand</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
            <span class="n">combined_chrom</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">chrom</span><span class="p">,</span> <span class="n">mapped_strand</span> <span class="ow">in</span> <span class="n">chrom_strand</span><span class="p">:</span>

                <span class="n">chrom</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">chrom</span><span class="p">)</span>
                <span class="n">chrom_strand</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="si">{}</span><span class="s1">__</span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">chrom</span><span class="p">,</span> <span class="n">mapped_strand</span><span class="p">)</span>
                <span class="k">assert</span> <span class="n">mapped_strand</span> <span class="ow">in</span> <span class="p">[</span>
                    <span class="s1">&#39;pos&#39;</span><span class="p">,</span> <span class="s1">&#39;neg&#39;</span>
                <span class="p">],</span> <span class="s1">&#39;Found strand </span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">mapped_strand</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">mapped_strand</span> <span class="o">==</span> <span class="s1">&#39;pos&#39;</span><span class="p">:</span>
                    <span class="n">mapped_strand</span> <span class="o">=</span> <span class="s1">&#39;+&#39;</span>
                <span class="k">elif</span> <span class="n">mapped_strand</span> <span class="o">==</span> <span class="s1">&#39;neg&#39;</span><span class="p">:</span>
                    <span class="n">mapped_strand</span> <span class="o">=</span> <span class="s1">&#39;-&#39;</span>

                <span class="n">chrom_obj</span> <span class="o">=</span> <span class="n">orientation_group</span><span class="p">[</span><span class="n">chrom_strand</span><span class="p">]</span>
                <span class="n">starts</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">chrom_obj</span><span class="p">[</span><span class="s1">&#39;positions&#39;</span><span class="p">])</span>
                <span class="n">ends</span> <span class="o">=</span> <span class="n">starts</span> <span class="o">+</span> <span class="mi">1</span>
                <span class="n">values</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">chrom_obj</span><span class="p">[</span><span class="s1">&#39;counts&#39;</span><span class="p">])</span>
                <span class="n">combined_chrom</span><span class="p">[</span><span class="n">chrom</span><span class="p">]</span> <span class="o">=</span> <span class="n">combined_chrom</span><span class="p">[</span><span class="n">chrom</span><span class="p">]</span><span class="o">.</span><span class="n">add</span><span class="p">(</span>
                    <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="n">values</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="n">starts</span><span class="p">),</span> <span class="n">fill_value</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>

                <span class="n">gene_strands</span> <span class="o">=</span> <span class="n">chrom_obj</span><span class="p">[</span><span class="s1">&#39;gene_strand&#39;</span><span class="p">]</span>
                <span class="k">if</span> <span class="n">mapped_strand</span> <span class="o">==</span> <span class="s1">&#39;+&#39;</span><span class="p">:</span>
                    <span class="n">bws</span><span class="p">[</span><span class="s1">&#39;pos_bw&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">addEntries</span><span class="p">(</span>
                        <span class="p">[</span><span class="n">chrom</span><span class="p">]</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">starts</span><span class="p">),</span>
                        <span class="n">starts</span><span class="p">,</span>
                        <span class="n">ends</span><span class="o">=</span><span class="n">ends</span><span class="p">,</span>
                        <span class="n">values</span><span class="o">=</span><span class="n">values</span><span class="p">)</span>
                    <span class="k">if</span> <span class="n">output_normalized</span><span class="p">:</span>
                        <span class="n">bws</span><span class="p">[</span><span class="s1">&#39;pos_norm_bw&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">addEntries</span><span class="p">(</span>
                            <span class="p">[</span><span class="n">chrom</span><span class="p">]</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">starts</span><span class="p">),</span>
                            <span class="n">starts</span><span class="p">,</span>
                            <span class="n">ends</span><span class="o">=</span><span class="n">ends</span><span class="p">,</span>
                            <span class="n">values</span><span class="o">=</span><span class="n">_scale_value_to_rpm</span><span class="p">(</span><span class="n">values</span><span class="p">,</span> <span class="n">read_total</span><span class="p">))</span>

                <span class="k">if</span> <span class="n">mapped_strand</span> <span class="o">==</span> <span class="s1">&#39;-&#39;</span><span class="p">:</span>
                    <span class="n">bws</span><span class="p">[</span><span class="s1">&#39;neg_bw&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">addEntries</span><span class="p">(</span>
                        <span class="p">[</span><span class="n">chrom</span><span class="p">]</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">starts</span><span class="p">),</span>
                        <span class="n">starts</span><span class="p">,</span>
                        <span class="n">ends</span><span class="o">=</span><span class="n">ends</span><span class="p">,</span>
                        <span class="n">values</span><span class="o">=</span><span class="n">values</span><span class="p">)</span>
                    <span class="k">if</span> <span class="n">output_normalized</span><span class="p">:</span>
                        <span class="n">bws</span><span class="p">[</span><span class="s1">&#39;neg_norm_bw&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">addEntries</span><span class="p">(</span>
                            <span class="p">[</span><span class="n">chrom</span><span class="p">]</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">starts</span><span class="p">),</span>
                            <span class="n">starts</span><span class="p">,</span>
                            <span class="n">ends</span><span class="o">=</span><span class="n">ends</span><span class="p">,</span>
                            <span class="n">values</span><span class="o">=</span><span class="n">_scale_value_to_rpm</span><span class="p">(</span><span class="n">values</span><span class="p">,</span> <span class="n">read_total</span><span class="p">))</span>
                <span class="n">gene_strand</span> <span class="o">=</span> <span class="n">chrom_obj</span><span class="p">[</span><span class="s1">&#39;gene_strand&#39;</span><span class="p">]</span>
                <span class="k">if</span> <span class="n">protocol</span> <span class="o">==</span> <span class="s1">&#39;forward&#39;</span><span class="p">:</span>
                    <span class="c1"># Should_keep mapped_strand + and gene_strand =</span>
                    <span class="c1"># OR mapped_strand - and gene_strand -i</span>
                    <span class="k">for</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">gene_strand</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span>
                            <span class="n">starts</span><span class="p">,</span> <span class="n">ends</span><span class="p">,</span> <span class="n">gene_strands</span><span class="p">,</span> <span class="n">values</span><span class="p">):</span>
                        <span class="k">if</span> <span class="n">gene_strand</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;+&#39;</span><span class="p">,</span> <span class="s1">&#39;-&#39;</span><span class="p">]:</span>
                            <span class="k">continue</span>
                        <span class="k">if</span> <span class="n">mapped_strand</span> <span class="o">==</span> <span class="n">gene_strand</span><span class="p">:</span>
                            <span class="n">bws</span><span class="p">[</span><span class="s1">&#39;collapsed_bw&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">addEntries</span><span class="p">([</span><span class="n">chrom</span><span class="p">],</span> <span class="p">[</span><span class="n">start</span><span class="p">],</span>
                                                           <span class="n">ends</span><span class="o">=</span><span class="p">[</span><span class="n">end</span><span class="p">],</span>
                                                           <span class="n">values</span><span class="o">=</span><span class="p">[</span><span class="n">value</span><span class="p">])</span>

                <span class="k">elif</span> <span class="n">protocol</span> <span class="o">==</span> <span class="s1">&#39;reverse&#39;</span><span class="p">:</span>
                    <span class="c1"># Should keep reads mapped to + and gene_strand -</span>
                    <span class="c1"># OR mapped_strand - and gene_strand +</span>
                    <span class="k">for</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">gene_strand</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span>
                            <span class="n">starts</span><span class="p">,</span> <span class="n">ends</span><span class="p">,</span> <span class="n">gene_strands</span><span class="p">,</span> <span class="n">values</span><span class="p">):</span>
                        <span class="k">if</span> <span class="n">gene_strand</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;+&#39;</span><span class="p">,</span> <span class="s1">&#39;-&#39;</span><span class="p">]:</span>
                            <span class="k">continue</span>
                        <span class="k">if</span> <span class="n">mapped_strand</span> <span class="o">==</span> <span class="n">complementary_strand</span><span class="p">(</span><span class="n">gene_strand</span><span class="p">):</span>
                            <span class="n">bws</span><span class="p">[</span><span class="s1">&#39;collapsed_bw&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">addEntries</span><span class="p">([</span><span class="n">chrom</span><span class="p">],</span> <span class="p">[</span><span class="n">start</span><span class="p">],</span>
                                                           <span class="n">ends</span><span class="o">=</span><span class="p">[</span><span class="n">end</span><span class="p">],</span>
                                                           <span class="n">values</span><span class="o">=</span><span class="p">[</span><span class="n">value</span><span class="p">])</span>

                <span class="k">elif</span> <span class="n">protocol</span> <span class="o">==</span> <span class="s1">&#39;unstranded&#39;</span><span class="p">:</span>
                    <span class="c1"># Count everything</span>
                    <span class="c1"># Do we still use this??</span>
                    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;Unstranded protocol not supported yet&#39;</span><span class="p">)</span>

            <span class="k">for</span> <span class="n">chrom</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">combined_chrom</span><span class="o">.</span><span class="n">keys</span><span class="p">())):</span>
                <span class="n">values</span> <span class="o">=</span> <span class="n">combined_chrom</span><span class="p">[</span><span class="n">chrom</span><span class="p">]</span><span class="o">.</span><span class="n">sort_index</span><span class="p">()</span>
                <span class="n">starts</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">values</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>
                <span class="n">ends</span> <span class="o">=</span> <span class="n">starts</span> <span class="o">+</span> <span class="mi">1</span>
                <span class="n">values</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">values</span><span class="o">.</span><span class="n">values</span><span class="p">)</span>
                <span class="n">bws</span><span class="p">[</span><span class="s1">&#39;combined_bw&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">addEntries</span><span class="p">(</span>
                    <span class="p">[</span><span class="n">chrom</span><span class="p">]</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">values</span><span class="p">),</span> <span class="n">starts</span><span class="p">,</span> <span class="n">ends</span><span class="o">=</span><span class="n">ends</span><span class="p">,</span> <span class="n">values</span><span class="o">=</span><span class="n">values</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">output_normalized</span><span class="p">:</span>
                    <span class="n">bws</span><span class="p">[</span><span class="s1">&#39;combined_norm_bw&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">addEntries</span><span class="p">(</span>
                        <span class="p">[</span><span class="n">chrom</span><span class="p">]</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">starts</span><span class="p">),</span>
                        <span class="n">starts</span><span class="p">,</span>
                        <span class="n">ends</span><span class="o">=</span><span class="n">ends</span><span class="p">,</span>
                        <span class="n">values</span><span class="o">=</span><span class="n">_scale_value_to_rpm</span><span class="p">(</span><span class="n">values</span><span class="p">,</span> <span class="n">read_total</span><span class="p">))</span>
            <span class="k">for</span> <span class="n">key</span><span class="p">,</span> <span class="n">bw</span> <span class="ow">in</span> <span class="n">six</span><span class="o">.</span><span class="n">iteritems</span><span class="p">(</span><span class="n">bws</span><span class="p">):</span>
                <span class="n">bw</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
    <span class="n">hdf</span><span class="o">.</span><span class="n">close</span><span class="p">()</span></div>


<div class="viewcode-block" id="tsv_to_bigwig"><a class="viewcode-back" href="../../riboraptor.html#riboraptor.hdf_parser.tsv_to_bigwig">[docs]</a><span class="k">def</span> <span class="nf">tsv_to_bigwig</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">chrom_lengths</span><span class="p">,</span> <span class="n">prefix</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Convert tsv (created by bam-coverage) to bigwig</span>

<span class="sd">    We create multiple bigwigs, each separately for fragment length,</span>
<span class="sd">    and the strand.</span>

<span class="sd">    This will output the following files, where N represents the fragment length</span>

<span class="sd">    N.5prime.pos.bw</span>
<span class="sd">    N.3prime.pos.bw</span>

<span class="sd">    N.5prime.neg.bw</span>
<span class="sd">    N.3prime.neg.bw</span>


<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    df: string or pd.DataFrame</span>
<span class="sd">        Path to bam-coverage outputted tsvi</span>
<span class="sd">    chrom_lengths: list of tuples</span>
<span class="sd">                   A list of tuples with chromsome name and size</span>
<span class="sd">                    [(&#39;chr1&#39;, 10000), [&#39;chr2&#39;, 5000]]</span>
<span class="sd">    prefix: string</span>
<span class="sd">            Path to store the final tsv</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">chrom_lengths</span><span class="p">,</span> <span class="n">six</span><span class="o">.</span><span class="n">string_types</span><span class="p">):</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">chrom_lengths</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span>
                <span class="s1">&#39;Need either list of tuples or chrom.sizes file to proceed&#39;</span><span class="p">)</span>
        <span class="n">sizes</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">chrom_lengths</span><span class="p">)</span> <span class="k">as</span> <span class="n">fh</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">fh</span><span class="p">:</span>
                <span class="n">chrom</span><span class="p">,</span> <span class="n">size</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
                <span class="n">sizes</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="nb">str</span><span class="p">(</span><span class="n">chrom</span><span class="p">),</span> <span class="nb">int</span><span class="p">(</span><span class="n">size</span><span class="p">)))</span>
        <span class="n">chrom_lengths</span> <span class="o">=</span> <span class="n">sizes</span>
    <span class="n">chrom_lengths</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">sorted</span><span class="p">(</span><span class="n">chrom_lengths</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">]))</span>
    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">six</span><span class="o">.</span><span class="n">string_types</span><span class="p">):</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_table</span><span class="p">(</span><span class="n">df</span><span class="p">)</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
    <span class="n">df</span><span class="o">.</span><span class="n">count_pos_strand</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">count_pos_strand</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">float</span><span class="p">)</span>
    <span class="n">df</span><span class="o">.</span><span class="n">count_neg_strand</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">count_neg_strand</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">float</span><span class="p">)</span>
    <span class="n">df_grouped</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s1">&#39;read_length&#39;</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">read_length</span><span class="p">,</span> <span class="n">read_len_group</span> <span class="ow">in</span> <span class="n">df_grouped</span><span class="p">:</span>
        <span class="n">orientation_grouped</span> <span class="o">=</span> <span class="n">read_len_group</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s1">&#39;orientation&#39;</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">orientation</span><span class="p">,</span> <span class="n">orientation_group</span> <span class="ow">in</span> <span class="n">orientation_grouped</span><span class="p">:</span>
            <span class="n">orientation_group</span> <span class="o">=</span> <span class="n">orientation_group</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span>
                <span class="n">by</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;chrom&#39;</span><span class="p">,</span> <span class="s1">&#39;start&#39;</span><span class="p">,</span> <span class="s1">&#39;end&#39;</span><span class="p">])</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span>
                    <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;read_length&#39;</span><span class="p">,</span> <span class="s1">&#39;orientation&#39;</span><span class="p">])</span>
            <span class="n">orientation_group_pos</span> <span class="o">=</span> <span class="n">orientation_group</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span>
                <span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;count_pos_strand&#39;</span><span class="p">:</span> <span class="s1">&#39;score&#39;</span><span class="p">})</span>
            <span class="n">orientation_group_pos</span><span class="p">[</span><span class="s1">&#39;strand&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;+&#39;</span>
            <span class="n">orientation_group_pos</span> <span class="o">=</span> <span class="n">orientation_group_pos</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span>
                <span class="n">by</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;chrom&#39;</span><span class="p">,</span> <span class="s1">&#39;start&#39;</span><span class="p">,</span> <span class="s1">&#39;end&#39;</span><span class="p">])</span>
            <span class="n">orientation_group_neg</span> <span class="o">=</span> <span class="n">orientation_group</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span>
                <span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;count_neg_strand&#39;</span><span class="p">:</span> <span class="s1">&#39;score&#39;</span><span class="p">})</span>
            <span class="n">orientation_group_neg</span><span class="p">[</span><span class="s1">&#39;strand&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;-&#39;</span>
            <span class="n">orientation_group_neg</span> <span class="o">=</span> <span class="n">orientation_group_neg</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span>
                <span class="n">by</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;chrom&#39;</span><span class="p">,</span> <span class="s1">&#39;start&#39;</span><span class="p">,</span> <span class="s1">&#39;end&#39;</span><span class="p">])</span>
            <span class="n">_create_bigwig_from_bed</span><span class="p">(</span>
                <span class="n">orientation_group_pos</span><span class="p">,</span> <span class="n">chrom_lengths</span><span class="p">,</span> <span class="s1">&#39;</span><span class="si">{}</span><span class="s1">_</span><span class="si">{}</span><span class="s1">_</span><span class="si">{}</span><span class="s1">_</span><span class="si">{}</span><span class="s1">.bw&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                    <span class="n">prefix</span><span class="p">,</span> <span class="n">read_length</span><span class="p">,</span> <span class="n">orientation</span><span class="p">,</span> <span class="s1">&#39;pos&#39;</span><span class="p">))</span>
            <span class="n">_create_bigwig_from_bed</span><span class="p">(</span>
                <span class="n">orientation_group_neg</span><span class="p">,</span> <span class="n">chrom_lengths</span><span class="p">,</span> <span class="s1">&#39;</span><span class="si">{}</span><span class="s1">_</span><span class="si">{}</span><span class="s1">_</span><span class="si">{}</span><span class="s1">_</span><span class="si">{}</span><span class="s1">.bw&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                    <span class="n">prefix</span><span class="p">,</span> <span class="n">read_length</span><span class="p">,</span> <span class="n">orientation</span><span class="p">,</span> <span class="s1">&#39;neg&#39;</span><span class="p">))</span></div>


<span class="k">def</span> <span class="nf">_process_gene_group</span><span class="p">(</span><span class="n">data</span><span class="p">):</span>
    <span class="n">filepath</span><span class="p">,</span> <span class="n">gene_group</span><span class="p">,</span> <span class="n">chromosome_lengths</span><span class="p">,</span> <span class="n">offset_5p</span><span class="p">,</span> <span class="n">offset_3p</span><span class="p">,</span> <span class="n">orientation</span><span class="p">,</span> <span class="n">n_bases</span><span class="p">,</span> <span class="n">fragment_length</span> <span class="o">=</span> <span class="n">data</span>
    <span class="n">h5py_obj</span> <span class="o">=</span> <span class="n">HDFParser</span><span class="p">(</span><span class="n">filepath</span><span class="p">)</span>
    <span class="n">regions</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span>
        <span class="nb">zip</span><span class="p">(</span><span class="n">gene_group</span><span class="p">[</span><span class="s1">&#39;chrom&#39;</span><span class="p">],</span> <span class="n">gene_group</span><span class="p">[</span><span class="s1">&#39;start&#39;</span><span class="p">],</span> <span class="n">gene_group</span><span class="p">[</span><span class="s1">&#39;end&#39;</span><span class="p">],</span>
            <span class="n">gene_group</span><span class="p">[</span><span class="s1">&#39;strand&#39;</span><span class="p">]))</span>
    <span class="n">strand</span> <span class="o">=</span> <span class="n">regions</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">3</span><span class="p">]</span>
    <span class="n">intervals</span> <span class="o">=</span> <span class="p">[</span><span class="n">Interval</span><span class="p">(</span><span class="n">i</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">i</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">i</span><span class="p">[</span><span class="mi">2</span><span class="p">],</span> <span class="n">i</span><span class="p">[</span><span class="mi">3</span><span class="p">])</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">regions</span><span class="p">]</span>
    <span class="n">intervals_combined</span><span class="p">,</span> <span class="n">gene_offset_5p</span><span class="p">,</span> <span class="n">gene_offset_3p</span> <span class="o">=</span> <span class="n">merge_intervals</span><span class="p">(</span>
        <span class="n">intervals</span><span class="p">,</span> <span class="n">chromosome_lengths</span><span class="p">,</span> <span class="n">offset_5p</span><span class="p">,</span> <span class="n">offset_3p</span><span class="p">)</span>
    <span class="n">regions</span> <span class="o">=</span> <span class="p">[(</span><span class="n">interval</span><span class="o">.</span><span class="n">chrom</span><span class="p">,</span> <span class="n">interval</span><span class="o">.</span><span class="n">start</span><span class="p">,</span> <span class="n">interval</span><span class="o">.</span><span class="n">end</span><span class="p">,</span> <span class="n">interval</span><span class="o">.</span><span class="n">strand</span><span class="p">)</span>
               <span class="k">for</span> <span class="n">interval</span> <span class="ow">in</span> <span class="n">intervals</span><span class="p">]</span>
    <span class="n">all_dfs</span> <span class="o">=</span> <span class="p">[</span>
        <span class="n">h5py_obj</span><span class="o">.</span><span class="n">get_coverage</span><span class="p">(</span><span class="n">region</span><span class="p">,</span> <span class="n">fragment_length</span><span class="p">,</span> <span class="n">orientation</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">region</span> <span class="ow">in</span> <span class="n">regions</span>
    <span class="p">]</span>
    <span class="n">h5py_obj</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
    <span class="n">all_dfs</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span><span class="n">all_dfs</span><span class="p">,</span> <span class="n">ignore_index</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="c1"># Reindex if the start position is &#39;-&#39;</span>
    <span class="k">if</span> <span class="n">strand</span> <span class="o">==</span> <span class="s1">&#39;-&#39;</span><span class="p">:</span>
        <span class="n">all_dfs</span> <span class="o">=</span> <span class="n">all_dfs</span><span class="o">.</span><span class="n">reindex</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">all_dfs</span><span class="o">.</span><span class="n">index</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span><span class="p">,</span> <span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="o">-</span><span class="mi">1</span><span class="p">))</span>
    <span class="n">max_positions</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">all_dfs</span><span class="o">.</span><span class="n">index</span><span class="p">),</span> <span class="n">n_bases</span><span class="p">)</span>
    <span class="c1"># Remove the &#39;start&#39; column as we will</span>
    <span class="c1"># dill with index from now on</span>
    <span class="n">coverage</span> <span class="o">=</span> <span class="n">all_dfs</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="nb">list</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="n">max_positions</span><span class="p">)),</span> <span class="n">all_dfs</span><span class="o">.</span><span class="n">columns</span><span class="p">[</span><span class="mi">1</span><span class="p">:]]</span>
    <span class="k">return</span> <span class="n">coverage</span>


<span class="k">def</span> <span class="nf">_gene_group_to_tsv</span><span class="p">(</span><span class="n">data</span><span class="p">):</span>
    <span class="n">gene_group</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s1">&#39;gene_group&#39;</span><span class="p">]</span>
    <span class="n">offset_5p</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s1">&#39;offset_5p&#39;</span><span class="p">]</span>
    <span class="n">offset_3p</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s1">&#39;offset_3p&#39;</span><span class="p">]</span>
    <span class="c1">#fragment_lengh = data[&#39;fragment_length&#39;]</span>
    <span class="c1">#orientation = data[&#39;orientation&#39;]</span>
    <span class="n">chromosome_lengths</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s1">&#39;chromosome_lengths&#39;</span><span class="p">]</span>
    <span class="n">n_bases</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s1">&#39;n_bases&#39;</span><span class="p">]</span>
    <span class="n">tsv</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s1">&#39;tsv&#39;</span><span class="p">]</span>
    <span class="n">tsv</span> <span class="o">=</span> <span class="n">tsv</span><span class="o">.</span><span class="n">set_index</span><span class="p">([</span><span class="s1">&#39;chrom&#39;</span><span class="p">,</span> <span class="s1">&#39;start&#39;</span><span class="p">,</span> <span class="s1">&#39;end&#39;</span><span class="p">])</span>
    <span class="n">regions</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span>
        <span class="nb">zip</span><span class="p">(</span><span class="n">gene_group</span><span class="p">[</span><span class="s1">&#39;chrom&#39;</span><span class="p">],</span> <span class="n">gene_group</span><span class="p">[</span><span class="s1">&#39;start&#39;</span><span class="p">],</span> <span class="n">gene_group</span><span class="p">[</span><span class="s1">&#39;end&#39;</span><span class="p">],</span>
            <span class="n">gene_group</span><span class="p">[</span><span class="s1">&#39;strand&#39;</span><span class="p">]))</span>
    <span class="n">strand</span> <span class="o">=</span> <span class="n">regions</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">3</span><span class="p">]</span>
    <span class="n">intervals</span> <span class="o">=</span> <span class="p">[</span><span class="n">Interval</span><span class="p">(</span><span class="n">i</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">i</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">i</span><span class="p">[</span><span class="mi">2</span><span class="p">],</span> <span class="n">i</span><span class="p">[</span><span class="mi">3</span><span class="p">])</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">regions</span><span class="p">]</span>
    <span class="n">intervals_combined</span><span class="p">,</span> <span class="n">gene_offset_5p</span><span class="p">,</span> <span class="n">gene_offset_3p</span> <span class="o">=</span> <span class="n">merge_intervals</span><span class="p">(</span>
        <span class="n">intervals</span><span class="p">,</span> <span class="n">chromosome_lengths</span><span class="p">,</span> <span class="n">offset_5p</span><span class="p">,</span> <span class="n">offset_3p</span><span class="p">)</span>
    <span class="n">regions</span> <span class="o">=</span> <span class="p">[(</span><span class="n">interval</span><span class="o">.</span><span class="n">chrom</span><span class="p">,</span> <span class="n">interval</span><span class="o">.</span><span class="n">start</span><span class="p">,</span> <span class="n">interval</span><span class="o">.</span><span class="n">end</span><span class="p">,</span> <span class="n">interval</span><span class="o">.</span><span class="n">strand</span><span class="p">)</span>
               <span class="k">for</span> <span class="n">interval</span> <span class="ow">in</span> <span class="n">intervals</span><span class="p">]</span>
    <span class="n">df</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">should_stop</span> <span class="o">=</span> <span class="kc">False</span>
    <span class="k">if</span> <span class="n">strand</span> <span class="o">==</span> <span class="s1">&#39;+&#39;</span><span class="p">:</span>
        <span class="n">total</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">for</span> <span class="n">region</span> <span class="ow">in</span> <span class="n">intervals_combined</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">should_stop</span><span class="p">:</span>
                <span class="k">break</span>
            <span class="k">for</span> <span class="n">position</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">region</span><span class="o">.</span><span class="n">start</span><span class="p">,</span> <span class="n">region</span><span class="o">.</span><span class="n">end</span><span class="p">):</span>
                <span class="k">if</span> <span class="n">should_stop</span><span class="p">:</span>
                    <span class="k">break</span>
                <span class="c1">#coverage_records =  tsv.loc[(region.chrom, position)]</span>
                <span class="n">df</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                    <span class="p">[</span><span class="n">region</span><span class="o">.</span><span class="n">chrom</span><span class="p">,</span> <span class="n">position</span><span class="p">,</span> <span class="n">position</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span> <span class="n">region</span><span class="o">.</span><span class="n">strand</span><span class="p">])</span>
                <span class="n">total</span> <span class="o">+=</span> <span class="mi">1</span>
                <span class="k">if</span> <span class="n">total</span> <span class="o">&gt;</span> <span class="n">n_bases</span><span class="p">:</span>
                    <span class="n">should_stop</span> <span class="o">=</span> <span class="kc">True</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="c1"># Iterate over the intervals in a reversed manner</span>
        <span class="k">for</span> <span class="n">region</span> <span class="ow">in</span> <span class="n">intervals_combined</span><span class="p">[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]:</span>
            <span class="k">if</span> <span class="n">should_stop</span><span class="p">:</span>
                <span class="k">break</span>
            <span class="k">for</span> <span class="n">position</span> <span class="ow">in</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">region</span><span class="o">.</span><span class="n">end</span> <span class="o">-</span> <span class="mi">1</span><span class="p">,</span> <span class="n">region</span><span class="o">.</span><span class="n">end</span> <span class="o">-</span> <span class="mi">1</span><span class="p">,</span> <span class="o">-</span><span class="mi">1</span><span class="p">):</span>
                <span class="k">if</span> <span class="n">should_stop</span><span class="p">:</span>
                    <span class="k">break</span>
                <span class="c1">#coverage_records =  tsv.loc[(region.chrom, position)]</span>
                <span class="n">df</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                    <span class="p">[</span><span class="n">region</span><span class="o">.</span><span class="n">chrom</span><span class="p">,</span> <span class="n">position</span><span class="p">,</span> <span class="n">position</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span> <span class="n">region</span><span class="o">.</span><span class="n">strand</span><span class="p">])</span>
                <span class="n">total</span> <span class="o">+=</span> <span class="mi">1</span>
                <span class="k">if</span> <span class="n">total</span> <span class="o">&gt;</span> <span class="n">n_bases</span><span class="p">:</span>
                    <span class="n">should_stop</span> <span class="o">=</span> <span class="kc">True</span>

    <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">df</span><span class="p">)</span>
    <span class="n">df</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;chrom&#39;</span><span class="p">,</span> <span class="s1">&#39;start&#39;</span><span class="p">,</span> <span class="s1">&#39;end&#39;</span><span class="p">,</span> <span class="s1">&#39;strand&#39;</span><span class="p">]</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">set_index</span><span class="p">([</span><span class="s1">&#39;chrom&#39;</span><span class="p">,</span> <span class="s1">&#39;start&#39;</span><span class="p">,</span> <span class="s1">&#39;end&#39;</span><span class="p">])</span>
    <span class="n">merged</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">tsv</span><span class="p">)</span>
    <span class="n">merged</span> <span class="o">=</span> <span class="n">merged</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>
    <span class="n">merged</span> <span class="o">=</span> <span class="n">merged</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;chrom&#39;</span><span class="p">,</span> <span class="s1">&#39;start&#39;</span><span class="p">,</span> <span class="s1">&#39;end&#39;</span><span class="p">])</span>
    <span class="k">if</span> <span class="n">strand</span> <span class="o">==</span> <span class="s1">&#39;+&#39;</span><span class="p">:</span>
        <span class="n">merged</span><span class="p">[</span><span class="s1">&#39;index&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">merged</span><span class="o">.</span><span class="n">index</span><span class="p">))</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">merged</span><span class="p">[</span><span class="s1">&#39;index&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">merged</span><span class="o">.</span><span class="n">index</span><span class="p">),</span> <span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="o">-</span><span class="mi">1</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">merged</span>


<div class="viewcode-block" id="create_metagene_from_multi_bigwig"><a class="viewcode-back" href="../../riboraptor.html#riboraptor.hdf_parser.create_metagene_from_multi_bigwig">[docs]</a><span class="k">def</span> <span class="nf">create_metagene_from_multi_bigwig</span><span class="p">(</span><span class="n">bed</span><span class="p">,</span>
                                      <span class="n">bigwigs</span><span class="p">,</span>
                                      <span class="n">max_positions</span><span class="o">=</span><span class="mi">1000</span><span class="p">,</span>
                                      <span class="n">offset_5p</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
                                      <span class="n">offset_3p</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
                                      <span class="n">n_jobs</span><span class="o">=</span><span class="mi">16</span><span class="p">,</span>
                                      <span class="n">saveto</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Collapse multiple bigwigs to get bigwig.</span>

<span class="sd">    Test case -&gt; we should be able to get the same coverage file as from export_metagene coverage</span>
<span class="sd">    if everything above this works okay.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">max_positions</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">warnings</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span><span class="s2">&quot;The max_positions is not specified, it could&quot;</span>
                      <span class="s2">&quot;take long time to calculate metagene coverage&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">max_positions</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">max_positions</span> <span class="o">&lt;=</span> <span class="mi">0</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span><span class="s1">&#39;The max_positions must be positive&#39;</span><span class="p">)</span>
        <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">bed</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;_&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="ow">in</span> <span class="n">__GENOMES_DB__</span><span class="p">:</span>
        <span class="n">splitted</span> <span class="o">=</span> <span class="n">bed</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;_&#39;</span><span class="p">)</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">splitted</span><span class="p">)</span> <span class="o">==</span> <span class="mi">2</span><span class="p">:</span>
            <span class="n">genome</span><span class="p">,</span> <span class="n">region_type</span> <span class="o">=</span> <span class="n">splitted</span>
        <span class="k">elif</span> <span class="nb">len</span><span class="p">(</span><span class="n">splitted</span><span class="p">)</span> <span class="o">==</span> <span class="mi">3</span><span class="p">:</span>
            <span class="n">genome</span> <span class="o">=</span> <span class="n">splitted</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
            <span class="n">region_type</span> <span class="o">=</span> <span class="p">(</span><span class="s1">&#39;_&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">splitted</span><span class="p">[</span><span class="mi">1</span><span class="p">:])</span>
        <span class="n">bed</span> <span class="o">=</span> <span class="n">_get_bed</span><span class="p">(</span><span class="n">region_type</span><span class="p">,</span> <span class="n">genome</span><span class="p">)</span>
    <span class="n">bed_df</span> <span class="o">=</span> <span class="n">pybedtools</span><span class="o">.</span><span class="n">BedTool</span><span class="p">(</span><span class="n">bed</span><span class="p">)</span><span class="o">.</span><span class="n">sort</span><span class="p">()</span><span class="o">.</span><span class="n">to_dataframe</span><span class="p">()</span>
    <span class="n">bed_df</span><span class="p">[</span><span class="s1">&#39;chrom&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">bed_df</span><span class="p">[</span><span class="s1">&#39;chrom&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>
    <span class="n">bed_df</span><span class="p">[</span><span class="s1">&#39;name&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">bed_df</span><span class="p">[</span><span class="s1">&#39;name&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>
    <span class="n">bed_grouped</span> <span class="o">=</span> <span class="n">bed_df</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s1">&#39;name&#39;</span><span class="p">)</span>

    <span class="n">position_counter</span> <span class="o">=</span> <span class="n">Counter</span><span class="p">()</span>
    <span class="n">metagene_coverage</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">()</span>
    <span class="k">for</span> <span class="n">bw</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span><span class="n">bigwigs</span><span class="p">):</span>
        <span class="n">data</span> <span class="o">=</span> <span class="p">[(</span><span class="n">gene_group</span><span class="p">,</span> <span class="n">bw</span><span class="o">.</span><span class="n">wig_location</span><span class="p">,</span> <span class="n">offset_5p</span><span class="p">,</span> <span class="n">offset_3p</span><span class="p">,</span>
                 <span class="n">max_positions</span><span class="p">)</span> <span class="k">for</span> <span class="n">gene_name</span><span class="p">,</span> <span class="n">gene_group</span> <span class="ow">in</span> <span class="n">bed_grouped</span><span class="p">]</span>
        <span class="n">aprun</span> <span class="o">=</span> <span class="n">ParallelExecutor</span><span class="p">(</span><span class="n">n_jobs</span><span class="o">=</span><span class="n">n_jobs</span><span class="p">)</span>
        <span class="n">total</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">bed_grouped</span><span class="o">.</span><span class="n">groups</span><span class="p">)</span>
        <span class="n">all_coverages</span> <span class="o">=</span> <span class="n">aprun</span><span class="p">(</span><span class="n">total</span><span class="o">=</span><span class="n">total</span><span class="p">)(</span>
            <span class="n">delayed</span><span class="p">(</span><span class="n">_multiprocess_gene_coverage</span><span class="p">)(</span><span class="n">d</span><span class="p">)</span> <span class="k">for</span> <span class="n">d</span> <span class="ow">in</span> <span class="n">data</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">norm_cov</span> <span class="ow">in</span> <span class="n">all_coverages</span><span class="p">:</span>
            <span class="n">metagene_coverage</span> <span class="o">=</span> <span class="n">metagene_coverage</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">norm_cov</span><span class="p">,</span> <span class="n">fill_value</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
            <span class="n">position_counter</span> <span class="o">+=</span> <span class="n">Counter</span><span class="p">(</span><span class="n">norm_cov</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">tolist</span><span class="p">())</span>

        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">position_counter</span><span class="p">)</span> <span class="o">!=</span> <span class="nb">len</span><span class="p">(</span><span class="n">metagene_coverage</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span><span class="s1">&#39;Gene normalizaed counter mismatch&#39;</span><span class="p">)</span>
            <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
    <span class="n">position_counter</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="n">position_counter</span><span class="p">)</span>
    <span class="n">metagene_coverage</span> <span class="o">=</span> <span class="n">metagene_coverage</span><span class="o">.</span><span class="n">div</span><span class="p">(</span><span class="n">position_counter</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">saveto</span><span class="p">:</span>
        <span class="n">mkdir_p</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span><span class="n">saveto</span><span class="p">))</span>
        <span class="n">to_write</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span>
            <span class="s1">&#39;position&#39;</span><span class="p">:</span> <span class="n">metagene_coverage</span><span class="o">.</span><span class="n">index</span><span class="p">,</span>
            <span class="s1">&#39;count&#39;</span><span class="p">:</span> <span class="n">metagene_coverage</span><span class="o">.</span><span class="n">values</span>
        <span class="p">})</span>
        <span class="n">to_write</span> <span class="o">=</span> <span class="n">to_write</span><span class="p">[[</span><span class="s1">&#39;position&#39;</span><span class="p">,</span> <span class="s1">&#39;count&#39;</span><span class="p">]]</span>
        <span class="n">to_write</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">saveto</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="nb">str</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">),</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">metagene_coverage</span></div>


<div class="viewcode-block" id="merge_bigwigs"><a class="viewcode-back" href="../../riboraptor.html#riboraptor.hdf_parser.merge_bigwigs">[docs]</a><span class="k">def</span> <span class="nf">merge_bigwigs</span><span class="p">(</span><span class="n">bigwigs</span><span class="p">,</span> <span class="n">chrom_sizes</span><span class="p">,</span> <span class="n">saveto</span><span class="p">,</span> <span class="n">scale</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Merge multiple bigwigs into one.</span>

<span class="sd">    Note: This seems impossible doing it through pybigWig way</span>
<span class="sd">    and hence the dependency on ucsc-bigWigMerge</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    bigwigs: list</span>
<span class="sd">             List of path to bigwigs</span>
<span class="sd">    chrom_size: string</span>
<span class="sd">                Path to chromsome.sizes file</span>
<span class="sd">    saveto: string</span>
<span class="sd">            Path for outputting bigwig</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">bigwigs</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">,</span> <span class="s1">&#39;Need more bigwigs to merge&#39;</span>
    <span class="k">with</span> <span class="n">tempfile</span><span class="o">.</span><span class="n">TemporaryDirectory</span><span class="p">(</span><span class="nb">dir</span><span class="o">=</span><span class="n">TMP_DIR_ROOT</span><span class="p">)</span> <span class="k">as</span> <span class="n">temp_dir</span><span class="p">:</span>
        <span class="n">bedgraph</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">temp_dir</span><span class="p">,</span> <span class="s1">&#39;</span><span class="si">{}</span><span class="s1">.bedGraph&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
            <span class="n">path_leaf</span><span class="p">(</span><span class="n">saveto</span><span class="p">)))</span>
        <span class="n">cmds</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;bigWigMerge&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="n">bigwigs</span> <span class="o">+</span> <span class="p">[</span><span class="n">bedgraph</span><span class="p">]</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">p</span> <span class="o">=</span> <span class="n">subprocess</span><span class="o">.</span><span class="n">Popen</span><span class="p">(</span>
                <span class="n">cmds</span><span class="p">,</span>
                <span class="n">stdout</span><span class="o">=</span><span class="n">subprocess</span><span class="o">.</span><span class="n">PIPE</span><span class="p">,</span>
                <span class="n">stderr</span><span class="o">=</span><span class="n">subprocess</span><span class="o">.</span><span class="n">PIPE</span><span class="p">,</span>
                <span class="n">universal_newlines</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
            <span class="n">stdout</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">communicate</span><span class="p">()</span>
        <span class="k">except</span> <span class="ne">FileNotFoundError</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span>
                <span class="s2">&quot;bigWigMerge not found on the path. This is an external &quot;</span>
                <span class="s2">&quot;tool from UCSC which can be downloaded from &quot;</span>
                <span class="s2">&quot;http://hgdownload.soe.ucsc.edu/admin/exe/. Alternatatively, use &quot;</span>
                <span class="s2">&quot;`conda install ucsc-bigwigmerge`&quot;</span><span class="p">)</span>
        <span class="n">cmds</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;bedSort&#39;</span><span class="p">,</span> <span class="n">bedgraph</span><span class="p">,</span> <span class="s1">&#39;</span><span class="si">{}</span><span class="s1">.sorted&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">bedgraph</span><span class="p">)]</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">p</span> <span class="o">=</span> <span class="n">subprocess</span><span class="o">.</span><span class="n">Popen</span><span class="p">(</span>
                <span class="n">cmds</span><span class="p">,</span>
                <span class="n">stdout</span><span class="o">=</span><span class="n">subprocess</span><span class="o">.</span><span class="n">PIPE</span><span class="p">,</span>
                <span class="n">stderr</span><span class="o">=</span><span class="n">subprocess</span><span class="o">.</span><span class="n">PIPE</span><span class="p">,</span>
                <span class="n">universal_newlines</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
            <span class="n">stdout</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">communicate</span><span class="p">()</span>
        <span class="k">except</span> <span class="ne">FileNotFoundError</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span>
                <span class="s2">&quot;bedSort not found on the path. This is an external &quot;</span>
                <span class="s2">&quot;tool from UCSC which can be downloaded from &quot;</span>
                <span class="s2">&quot;http://hgdownload.soe.ucsc.edu/admin/exe/. Alternatatively, use &quot;</span>
                <span class="s2">&quot;`conda install ucsc-bedsort`&quot;</span><span class="p">)</span>

        <span class="n">cmds</span> <span class="o">=</span> <span class="p">[</span>
            <span class="s1">&#39;bedGraphToBigWig&#39;</span><span class="p">,</span> <span class="s1">&#39;</span><span class="si">{}</span><span class="s1">.sorted&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">bedgraph</span><span class="p">),</span> <span class="n">chrom_sizes</span><span class="p">,</span>
            <span class="n">saveto</span>
        <span class="p">]</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">p</span> <span class="o">=</span> <span class="n">subprocess</span><span class="o">.</span><span class="n">Popen</span><span class="p">(</span>
                <span class="n">cmds</span><span class="p">,</span>
                <span class="n">stdout</span><span class="o">=</span><span class="n">subprocess</span><span class="o">.</span><span class="n">PIPE</span><span class="p">,</span>
                <span class="n">stderr</span><span class="o">=</span><span class="n">subprocess</span><span class="o">.</span><span class="n">PIPE</span><span class="p">,</span>
                <span class="n">universal_newlines</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
            <span class="n">stdout</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">communicate</span><span class="p">()</span>
        <span class="k">except</span> <span class="ne">FileNotFoundError</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span>
                <span class="s2">&quot;bedGraphToBigwig not found on the path. This is an external &quot;</span>
                <span class="s2">&quot;tool from UCSC which can be downloaded from &quot;</span>
                <span class="s2">&quot;http://hgdownload.soe.ucsc.edu/admin/exe/. Alternatatively, use &quot;</span>
                <span class="s2">&quot;`conda install ucsc-bedgraphtobigwig`&quot;</span><span class="p">)</span></div>


<div class="viewcode-block" id="HDFParser"><a class="viewcode-back" href="../../riboraptor.html#riboraptor.hdf_parser.HDFParser">[docs]</a><span class="k">class</span> <span class="nc">HDFParser</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>
    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filepath</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">filepath</span> <span class="o">=</span> <span class="n">filepath</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">h5py_obj</span> <span class="o">=</span> <span class="n">h5py</span><span class="o">.</span><span class="n">File</span><span class="p">(</span><span class="n">filepath</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span>
        <span class="n">chrom_names</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">h5py_obj</span><span class="p">[</span><span class="s1">&#39;chrom_names&#39;</span><span class="p">]</span>
        <span class="n">chrom_sizes</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">h5py_obj</span><span class="p">[</span><span class="s1">&#39;chrom_sizes&#39;</span><span class="p">]</span>
        <span class="n">chrom_lengths</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span>
            <span class="p">[(</span><span class="n">chrom</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;_neg&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;_pos&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">),</span> <span class="nb">int</span><span class="p">(</span><span class="n">size</span><span class="p">))</span>
             <span class="k">for</span> <span class="n">chrom</span><span class="p">,</span> <span class="n">size</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">chrom_names</span><span class="p">,</span> <span class="n">chrom_sizes</span><span class="p">)])</span>
        <span class="n">chrom_lengths</span> <span class="o">=</span> <span class="n">OrderedDict</span><span class="p">(</span><span class="nb">sorted</span><span class="p">(</span><span class="n">chrom_lengths</span><span class="o">.</span><span class="n">items</span><span class="p">()))</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">chromosome_lengths</span> <span class="o">=</span> <span class="n">chrom_lengths</span>

<div class="viewcode-block" id="HDFParser.close"><a class="viewcode-back" href="../../riboraptor.html#riboraptor.hdf_parser.HDFParser.close">[docs]</a>    <span class="k">def</span> <span class="nf">close</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">h5py_obj</span><span class="o">.</span><span class="n">close</span><span class="p">()</span></div>

    <span class="k">def</span> <span class="nf">__enter__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="bp">self</span>

    <span class="k">def</span> <span class="nf">__exit__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">exc_type</span><span class="p">,</span> <span class="n">exc_value</span><span class="p">,</span> <span class="n">traceback</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">h5py_obj</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

<div class="viewcode-block" id="HDFParser.get_coverage"><a class="viewcode-back" href="../../riboraptor.html#riboraptor.hdf_parser.HDFParser.get_coverage">[docs]</a>    <span class="k">def</span> <span class="nf">get_coverage</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span>
                     <span class="n">region</span><span class="o">=</span><span class="p">(</span><span class="kc">None</span><span class="p">,</span> <span class="kc">None</span><span class="p">,</span> <span class="kc">None</span><span class="p">,</span> <span class="kc">None</span><span class="p">),</span>
                     <span class="n">fragment_length</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                     <span class="n">orientation</span><span class="o">=</span><span class="s1">&#39;5prime&#39;</span><span class="p">,</span>
                     <span class="n">outprefix</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Get coverage for a selected region.</span>

<span class="sd">        Since the region is user provided, we assume the user is aware if it is strand</span>
<span class="sd">        specfic or not</span>

<span class="sd">        Paramters</span>
<span class="sd">        ---------</span>
<span class="sd">        region: tuple</span>
<span class="sd">                (chromosome, start[0-based], end[1-based], strand(+/-))</span>
<span class="sd">        fragment_length: list or string</span>
<span class="sd">                         list of fragment lengths to use</span>
<span class="sd">                         Example: [20, 21, 22] or &#39;all&#39;</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">chrom</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">stop</span><span class="p">,</span> <span class="n">strand</span> <span class="o">=</span> <span class="n">region</span>
        <span class="n">start</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">start</span><span class="p">)</span>
        <span class="n">stop</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">stop</span><span class="p">)</span>
        <span class="k">assert</span> <span class="n">chrom</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">,</span> <span class="s1">&#39;chromosome not set&#39;</span>
        <span class="k">assert</span> <span class="n">start</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">,</span> <span class="s1">&#39;start position is not set&#39;</span>
        <span class="k">assert</span> <span class="n">stop</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">,</span> <span class="s1">&#39;end position not set&#39;</span>
        <span class="k">assert</span> <span class="n">strand</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">,</span> <span class="s1">&#39;strand is not set&#39;</span>
        <span class="k">assert</span> <span class="n">start</span> <span class="o">&lt;</span> <span class="n">stop</span><span class="p">,</span> <span class="s1">&#39;start should be &lt; stop&#39;</span>
        <span class="k">assert</span> <span class="n">fragment_length</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">,</span> <span class="s1">&#39;fragment_length not set&#39;</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">fragment_length</span><span class="p">,</span> <span class="nb">int</span><span class="p">):</span>
            <span class="n">fragment_length</span> <span class="o">=</span> <span class="p">[</span><span class="n">fragment_length</span><span class="p">]</span>

        <span class="k">if</span> <span class="n">fragment_length</span> <span class="o">==</span> <span class="s1">&#39;all&#39;</span><span class="p">:</span>
            <span class="n">h5py_fragments</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">h5py_obj</span><span class="p">[</span><span class="s1">&#39;read_lengths&#39;</span><span class="p">]</span>
            <span class="n">fragment_length</span> <span class="o">=</span> <span class="n">h5py_fragments</span>
        <span class="n">fragment_length</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">map</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="nb">str</span><span class="p">(</span><span class="n">x</span><span class="p">),</span> <span class="n">fragment_length</span><span class="p">))</span>
        <span class="n">coverages_normalized</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()</span>
        <span class="n">coverages</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()</span>
        <span class="n">position_counter</span> <span class="o">=</span> <span class="n">Counter</span><span class="p">()</span>
        <span class="k">for</span> <span class="n">l</span> <span class="ow">in</span> <span class="n">fragment_length</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">strand</span> <span class="o">==</span> <span class="s1">&#39;-&#39;</span><span class="p">:</span>
                <span class="n">chrom_name</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="si">{}</span><span class="s1">__neg&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">chrom</span><span class="p">)</span>
            <span class="k">elif</span> <span class="n">strand</span> <span class="o">==</span> <span class="s1">&#39;+&#39;</span><span class="p">:</span>
                <span class="n">chrom_name</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="si">{}</span><span class="s1">__pos&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">chrom</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;strand ill-defined: </span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">strand</span><span class="p">))</span>
            <span class="n">root_obj</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">h5py_obj</span><span class="p">[</span><span class="s1">&#39;fragments&#39;</span><span class="p">][</span><span class="n">l</span><span class="p">][</span><span class="n">orientation</span><span class="p">]</span>
            <span class="k">if</span> <span class="n">chrom_name</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">root_obj</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
                <span class="c1"># This chromosome does not exist in the</span>
                <span class="c1"># key value store</span>
                <span class="c1"># So should returns zeros all way</span>
                <span class="n">coverage</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span>
                    <span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">*</span> <span class="p">(</span><span class="n">stop</span> <span class="o">-</span> <span class="n">start</span><span class="p">),</span> <span class="n">index</span><span class="o">=</span><span class="nb">range</span><span class="p">(</span><span class="n">start</span><span class="p">,</span> <span class="n">stop</span><span class="p">))</span>

            <span class="k">else</span><span class="p">:</span>
                <span class="n">chrom_obj</span> <span class="o">=</span> <span class="n">root_obj</span><span class="p">[</span><span class="n">chrom_name</span><span class="p">]</span>
                <span class="n">counts_series</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span>
                    <span class="nb">list</span><span class="p">(</span><span class="n">chrom_obj</span><span class="p">[</span><span class="s1">&#39;counts&#39;</span><span class="p">]),</span>
                    <span class="n">index</span><span class="o">=</span><span class="nb">list</span><span class="p">(</span><span class="n">chrom_obj</span><span class="p">[</span><span class="s1">&#39;positions&#39;</span><span class="p">]))</span>
                <span class="c1">#print(counts_series)</span>
                <span class="n">coverage</span> <span class="o">=</span> <span class="n">counts_series</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="n">start</span><span class="p">,</span> <span class="n">stop</span><span class="p">)))</span>
                <span class="k">if</span> <span class="n">coverage</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                    <span class="n">coverage</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span>
                        <span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">*</span> <span class="p">(</span><span class="n">stop</span> <span class="o">-</span> <span class="n">start</span><span class="p">),</span> <span class="n">index</span><span class="o">=</span><span class="nb">range</span><span class="p">(</span><span class="n">start</span><span class="p">,</span> <span class="n">stop</span><span class="p">))</span>
                <span class="n">coverage</span> <span class="o">=</span> <span class="n">coverage</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
            <span class="c1"># Mean is taken by summing the rows</span>
            <span class="n">coverage_mean</span> <span class="o">=</span> <span class="n">coverage</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">skipna</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
            <span class="c1"># to normalize</span>
            <span class="c1"># we divide the sum by vector obtained in previous</span>
            <span class="c1"># such that each column gets divided</span>
            <span class="n">coverage_normalized</span> <span class="o">=</span> <span class="n">coverage</span><span class="o">.</span><span class="n">divide</span><span class="p">(</span><span class="n">coverage_mean</span><span class="p">)</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
            <span class="n">coverages</span> <span class="o">=</span> <span class="n">coverages</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
                <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">coverage</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="n">l</span><span class="p">)]),</span> <span class="n">how</span><span class="o">=</span><span class="s1">&#39;outer&#39;</span><span class="p">)</span>
            <span class="n">coverages_normalized</span> <span class="o">=</span> <span class="n">coverages_normalized</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
                <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">coverage_normalized</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="n">l</span><span class="p">)]),</span>
                <span class="n">how</span><span class="o">=</span><span class="s1">&#39;outer&#39;</span><span class="p">)</span>
            <span class="n">position_counter</span> <span class="o">+=</span> <span class="n">Counter</span><span class="p">(</span><span class="n">coverage</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">tolist</span><span class="p">())</span>
        <span class="n">position_counter</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="n">Counter</span><span class="p">(</span><span class="n">position_counter</span><span class="p">))</span><span class="o">.</span><span class="n">sort_index</span><span class="p">()</span>
        <span class="n">coverage_sum</span> <span class="o">=</span> <span class="n">coverages</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
        <span class="n">coverage_normalized_sum</span> <span class="o">=</span> <span class="n">coverages_normalized</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
        <span class="n">coverage_sum_div</span> <span class="o">=</span> <span class="n">coverage_sum</span><span class="o">.</span><span class="n">divide</span><span class="p">(</span><span class="n">position_counter</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="s1">&#39;index&#39;</span><span class="p">)</span>
        <span class="n">coverage_normalized_sum_div</span> <span class="o">=</span> <span class="n">coverage_normalized_sum</span><span class="o">.</span><span class="n">divide</span><span class="p">(</span>
            <span class="n">position_counter</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="s1">&#39;index&#39;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">outprefix</span><span class="p">:</span>
            <span class="n">mkdir_p</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span><span class="n">outprefix</span><span class="p">))</span>
            <span class="n">coverage_sum_div</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span>
                <span class="s1">&#39;</span><span class="si">{}</span><span class="s1">_raw.tsv&#39;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
            <span class="n">coverage_normalized_sum_div</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span>
                <span class="s1">&#39;</span><span class="si">{}</span><span class="s1">_normalized.tsv&#39;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>

        <span class="n">coverages</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="s1">&#39;start&#39;</span>
        <span class="n">coverages</span> <span class="o">=</span> <span class="n">coverages</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>
        <span class="n">coverages_normalized</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="s1">&#39;start&#39;</span>
        <span class="n">coverages_normalized</span> <span class="o">=</span> <span class="n">coverages_normalized</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>
        <span class="k">return</span> <span class="n">coverages</span><span class="p">,</span> <span class="n">coverages_normalized</span><span class="p">,</span> <span class="n">coverage_sum_div</span><span class="p">,</span> <span class="n">coverage_normalized_sum_div</span></div>

<div class="viewcode-block" id="HDFParser.get_read_length_dist"><a class="viewcode-back" href="../../riboraptor.html#riboraptor.hdf_parser.HDFParser.get_read_length_dist">[docs]</a>    <span class="k">def</span> <span class="nf">get_read_length_dist</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="n">read_lengths</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span>
            <span class="nb">map</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="nb">int</span><span class="p">(</span><span class="n">x</span><span class="p">),</span> <span class="bp">self</span><span class="o">.</span><span class="n">h5py_obj</span><span class="p">[</span><span class="s1">&#39;read_lengths&#39;</span><span class="p">]))</span>
        <span class="n">read_counts</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">h5py_obj</span><span class="p">[</span><span class="s1">&#39;read_lengths_counts&#39;</span><span class="p">])</span>
        <span class="k">return</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="n">read_counts</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="n">read_lengths</span><span class="p">)</span><span class="o">.</span><span class="n">sort_index</span><span class="p">()</span></div>

<div class="viewcode-block" id="HDFParser.get_query_alignment_length_dist"><a class="viewcode-back" href="../../riboraptor.html#riboraptor.hdf_parser.HDFParser.get_query_alignment_length_dist">[docs]</a>    <span class="k">def</span> <span class="nf">get_query_alignment_length_dist</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="n">read_lengths</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span>
            <span class="nb">map</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="nb">int</span><span class="p">(</span><span class="n">x</span><span class="p">),</span> <span class="bp">self</span><span class="o">.</span><span class="n">h5py_obj</span><span class="p">[</span><span class="s1">&#39;query_alignment_lengths&#39;</span><span class="p">]))</span>
        <span class="n">read_counts</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">h5py_obj</span><span class="p">[</span><span class="s1">&#39;query_alignment_lengths_counts&#39;</span><span class="p">])</span>
        <span class="k">return</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="n">read_counts</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="n">read_lengths</span><span class="p">)</span><span class="o">.</span><span class="n">sort_index</span><span class="p">()</span></div></div>


<div class="viewcode-block" id="normalize_bw_hdf"><a class="viewcode-back" href="../../riboraptor.html#riboraptor.hdf_parser.normalize_bw_hdf">[docs]</a><span class="k">def</span> <span class="nf">normalize_bw_hdf</span><span class="p">(</span><span class="n">bw</span><span class="p">,</span> <span class="n">hdf</span><span class="p">,</span> <span class="n">read_length</span><span class="p">,</span> <span class="n">outbw</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Normalize a fragment specific bigwig to RPM for that fragment length</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    bw: string</span>
<span class="sd">        Path to bigwig file</span>
<span class="sd">    hdf: string</span>
<span class="sd">         Path to hdf file (used for fetching total counts for that read type)</span>
<span class="sd">    read_length: int</span>
<span class="sd">                 Fragment length</span>
<span class="sd">    outbw: string</span>
<span class="sd">           Path to output bigwig</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">hdf</span> <span class="o">=</span> <span class="n">HDFParser</span><span class="p">(</span><span class="n">hdf</span><span class="p">)</span>
    <span class="n">read_length_dist</span> <span class="o">=</span> <span class="n">hdf</span><span class="o">.</span><span class="n">get_read_length_dist</span><span class="p">()</span>
    <span class="n">chrom_sizes</span> <span class="o">=</span> <span class="n">hdf</span><span class="o">.</span><span class="n">chromosome_lengths</span>
    <span class="k">with</span> <span class="n">tempfile</span><span class="o">.</span><span class="n">TemporaryDirectory</span><span class="p">(</span><span class="nb">dir</span><span class="o">=</span><span class="n">TMP_DIR_ROOT</span><span class="p">)</span> <span class="k">as</span> <span class="n">temp_dir</span><span class="p">:</span>
        <span class="n">chrom_sizes_file</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">temp_dir</span><span class="p">,</span> <span class="s1">&#39;chrom.sizes&#39;</span><span class="p">)</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">chrom_sizes_file</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">fh</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">chrom</span><span class="p">,</span> <span class="n">size</span> <span class="ow">in</span> <span class="n">six</span><span class="o">.</span><span class="n">iteritems</span><span class="p">(</span><span class="n">chrom_sizes</span><span class="p">):</span>
                <span class="n">fh</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">{}</span><span class="se">\t</span><span class="si">{}</span><span class="se">\n</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">chrom</span><span class="p">,</span> <span class="n">size</span><span class="p">))</span>
        <span class="n">total</span> <span class="o">=</span> <span class="n">read_length_dist</span><span class="p">[</span><span class="n">read_length</span><span class="p">]</span>
        <span class="n">scale_factor</span> <span class="o">=</span> <span class="mi">1</span> <span class="o">/</span> <span class="p">(</span><span class="n">total</span> <span class="o">/</span> <span class="mf">1e6</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;total_reads: </span><span class="si">{}</span><span class="s1"> | scale_factor: </span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">total</span><span class="p">,</span> <span class="n">scale_factor</span><span class="p">))</span>
        <span class="n">scale_bigwig</span><span class="p">(</span><span class="n">bw</span><span class="p">,</span> <span class="n">chrom_sizes_file</span><span class="p">,</span> <span class="n">outbw</span><span class="p">,</span> <span class="n">scale_factor</span><span class="p">)</span></div>
</pre></div>

          </div>
            
        </div>
        <div class="clearfix"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="../../index.html">riboraptor 0.3.0 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="../index.html" >Module code</a> &#187;</li> 
      </ul>
    </div>
<script type="text/javascript">
  $("#mobile-toggle a").click(function () {
    $("#left-column").toggle();
  });
</script>
<script type="text/javascript" src="../../_static/js/bootstrap.js"></script>
  <div class="footer">
    &copy; Copyright 2018, Saket Choudhary. Created using <a href="http://sphinx.pocoo.org/">Sphinx</a>.
  </div>
  </body>
</html>