
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="X-UA-Compatible" content="IE=Edge" />
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
  
  <!-- Licensed under the Apache 2.0 License -->
  <link rel="stylesheet" type="text/css" href="../../_static/fonts/open-sans/stylesheet.css" />
  <!-- Licensed under the SIL Open Font License -->
  <link rel="stylesheet" type="text/css" href="../../_static/fonts/source-serif-pro/source-serif-pro.css" />
  <link rel="stylesheet" type="text/css" href="../../_static/css/bootstrap.min.css" />
  <link rel="stylesheet" type="text/css" href="../../_static/css/bootstrap-theme.min.css" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
    <title>riboraptor.infer_protocol &#8212; riboraptor 0.3.0 documentation</title>
    <link rel="stylesheet" href="../../_static/guzzle.css" type="text/css" />
    <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../../_static/graphviz.css" />
    <script type="text/javascript" id="documentation_options" data-url_root="../../" src="../../_static/documentation_options.js"></script>
    <script type="text/javascript" src="../../_static/jquery.js"></script>
    <script type="text/javascript" src="../../_static/underscore.js"></script>
    <script type="text/javascript" src="../../_static/doctools.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
  
   

  </head><body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="../../index.html">riboraptor 0.3.0 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="../index.html" accesskey="U">Module code</a> &#187;</li> 
      </ul>
    </div>
    <div class="container-wrapper">

      <div id="mobile-toggle">
        <a href="#"><span class="glyphicon glyphicon-align-justify" aria-hidden="true"></span></a>
      </div>
  <div id="left-column">
    <div class="sphinxsidebar">
<div class="sidebar-block">
  <div class="sidebar-wrapper">
    <h2>Contents</h2>
    <div class="sidebar-localtoc">
      <ul>
<li class="toctree-l1"><a class="reference internal" href="../../installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../example_workflow.html">Example Workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../cmd-manual.html">Manual</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../api-usage.html">API Usage</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../scores.html">Scores</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../modules.html">riboraptor</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../history.html">History</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../contributing.html">Contributing</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../authors.html">Credits</a></li>
</ul>

    </div>
  </div>
</div>
<div class="sidebar-block">
  <div class="sidebar-wrapper">
    <div id="main-search">
      <form class="form-inline" action="../../search.html" method="GET" role="form">
        <div class="input-group">
          <input name="q" type="text" class="form-control" placeholder="Search...">
        </div>
        <input type="hidden" name="check_keywords" value="yes" />
        <input type="hidden" name="area" value="default" />
      </form>
    </div>
  </div>
</div>
      
    </div>
  </div>
        <div id="right-column">
          
          <div role="navigation" aria-label="breadcrumbs navigation">
            <ol class="breadcrumb">
              <li><a href="../../index.html">Docs</a></li>
              
                <li><a href="../index.html">Module code</a></li>
              
              <li>riboraptor.infer_protocol</li>
            </ol>
          </div>
          
          <div class="document clearer body">
            
  <h1>Source code for riboraptor.infer_protocol</h1><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">__future__</span> <span class="k">import</span> <span class="n">absolute_import</span>
<span class="kn">from</span> <span class="nn">__future__</span> <span class="k">import</span> <span class="n">division</span>
<span class="kn">from</span> <span class="nn">__future__</span> <span class="k">import</span> <span class="n">print_function</span>
<span class="kn">from</span> <span class="nn">__future__</span> <span class="k">import</span> <span class="n">unicode_literals</span>

<span class="kn">import</span> <span class="nn">pysam</span>
<span class="kn">from</span> <span class="nn">.helpers</span> <span class="k">import</span> <span class="n">is_read_uniq_mapping</span><span class="p">,</span> <span class="n">create_bam_index</span>
<span class="kn">from</span> <span class="nn">.helpers</span> <span class="k">import</span> <span class="n">read_refseq_bed</span><span class="p">,</span> <span class="n">read_bed_as_intervaltree</span>
<span class="kn">from</span> <span class="nn">collections</span> <span class="k">import</span> <span class="n">Counter</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>


<div class="viewcode-block" id="infer_protocol"><a class="viewcode-back" href="../../riboraptor.html#riboraptor.infer_protocol.infer_protocol">[docs]</a><span class="k">def</span> <span class="nf">infer_protocol</span><span class="p">(</span><span class="n">bam</span><span class="p">,</span> <span class="n">bed</span><span class="p">,</span> <span class="n">n_reads</span><span class="o">=</span><span class="mi">500000</span><span class="p">,</span> <span class="n">drop_probability</span><span class="o">=</span><span class="mf">0.2</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Infer strandedness protocol given a bam file</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    bam: string</span>
<span class="sd">         Path to bam file</span>
<span class="sd">    bed: string</span>
<span class="sd">         Path to gene bed file</span>
<span class="sd">    n_reads: int</span>
<span class="sd">             Number of reads to use (downsampled)</span>
<span class="sd">    drop_probability: float [0-1]</span>
<span class="sd">                      Reads are randomly droppped with this probability to</span>
<span class="sd">                      simulate sampling</span>


<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    protocol: string</span>
<span class="sd">              unstranded/forward/reverse</span>
<span class="sd">    forward_mapped_reads: float</span>
<span class="sd">          Proportion of reads of type + mapping to + (++) or - mapping to - (--)</span>
<span class="sd">    reverse_mapped_reads: float</span>
<span class="sd">          Proportion of reads of type + mapping to - (+-) or - mapping to + (-+)</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">seed</span><span class="p">(</span><span class="mi">42</span><span class="p">)</span>
    <span class="n">iteration</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">create_bam_index</span><span class="p">(</span><span class="n">bam</span><span class="p">)</span>
    <span class="n">bam</span> <span class="o">=</span> <span class="n">pysam</span><span class="o">.</span><span class="n">AlignmentFile</span><span class="p">(</span><span class="n">bam</span><span class="p">,</span> <span class="s1">&#39;rb&#39;</span><span class="p">)</span>
    <span class="n">bed</span> <span class="o">=</span> <span class="n">read_bed_as_intervaltree</span><span class="p">(</span><span class="n">bed</span><span class="p">)</span>
    <span class="n">strandedness</span> <span class="o">=</span> <span class="n">Counter</span><span class="p">()</span>
    <span class="k">for</span> <span class="n">read</span> <span class="ow">in</span> <span class="n">bam</span><span class="o">.</span><span class="n">fetch</span><span class="p">():</span>
        <span class="k">if</span> <span class="n">drop_probability</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">should_drop</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span>
                <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">p</span><span class="o">=</span><span class="p">[</span><span class="n">drop_probability</span><span class="p">,</span> <span class="mi">1</span> <span class="o">-</span> <span class="n">drop_probability</span><span class="p">])</span>
            <span class="k">if</span> <span class="n">should_drop</span><span class="p">:</span>
                <span class="k">continue</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">is_read_uniq_mapping</span><span class="p">(</span><span class="n">read</span><span class="p">):</span>
            <span class="k">continue</span>
        <span class="k">if</span> <span class="n">read</span><span class="o">.</span><span class="n">is_reverse</span><span class="p">:</span>
            <span class="n">mapped_strand</span> <span class="o">=</span> <span class="s1">&#39;-&#39;</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">mapped_strand</span> <span class="o">=</span> <span class="s1">&#39;+&#39;</span>
        <span class="n">mapped_start</span> <span class="o">=</span> <span class="n">read</span><span class="o">.</span><span class="n">reference_start</span>
        <span class="n">mapped_end</span> <span class="o">=</span> <span class="n">read</span><span class="o">.</span><span class="n">reference_end</span>
        <span class="n">chrom</span> <span class="o">=</span> <span class="n">read</span><span class="o">.</span><span class="n">reference_name</span>
        <span class="n">gene_strand</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">bed</span><span class="p">[</span><span class="n">chrom</span><span class="p">]</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="n">mapped_start</span><span class="p">,</span> <span class="n">mapped_end</span><span class="p">)))</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">gene_strand</span><span class="p">)</span> <span class="o">!=</span> <span class="mi">1</span><span class="p">:</span>
            <span class="c1"># Filter out genes with ambiguous strand info</span>
            <span class="c1"># (those) that have a tx_start on opposite strands</span>
            <span class="k">continue</span>
        <span class="n">gene_strand</span> <span class="o">=</span> <span class="n">gene_strand</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">strandedness</span><span class="p">[</span><span class="s1">&#39;</span><span class="si">{}{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">mapped_strand</span><span class="p">,</span> <span class="n">gene_strand</span><span class="p">)]</span> <span class="o">+=</span> <span class="mi">1</span>
        <span class="n">iteration</span> <span class="o">+=</span> <span class="mi">1</span>
        <span class="k">if</span> <span class="n">iteration</span> <span class="o">&gt;=</span> <span class="n">n_reads</span><span class="p">:</span>
            <span class="k">break</span>
    <span class="c1">## Add pseudocounts</span>
    <span class="n">strandedness</span><span class="p">[</span><span class="s1">&#39;++&#39;</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>
    <span class="n">strandedness</span><span class="p">[</span><span class="s1">&#39;--&#39;</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>
    <span class="n">strandedness</span><span class="p">[</span><span class="s1">&#39;+-&#39;</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>
    <span class="n">strandedness</span><span class="p">[</span><span class="s1">&#39;-+&#39;</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>

    <span class="n">total</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="n">strandedness</span><span class="o">.</span><span class="n">values</span><span class="p">())</span>
    <span class="n">forward_mapped_reads</span> <span class="o">=</span> <span class="p">(</span><span class="n">strandedness</span><span class="p">[</span><span class="s1">&#39;++&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="n">strandedness</span><span class="p">[</span><span class="s1">&#39;--&#39;</span><span class="p">])</span> <span class="o">/</span> <span class="n">total</span>
    <span class="n">reverse_mapped_reads</span> <span class="o">=</span> <span class="p">(</span><span class="n">strandedness</span><span class="p">[</span><span class="s1">&#39;-+&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="n">strandedness</span><span class="p">[</span><span class="s1">&#39;+-&#39;</span><span class="p">])</span> <span class="o">/</span> <span class="n">total</span>
    <span class="n">ratio</span> <span class="o">=</span> <span class="n">forward_mapped_reads</span> <span class="o">/</span> <span class="n">reverse_mapped_reads</span>
    <span class="c1"># Prefer checking for unstrandedness</span>
    <span class="c1"># Check if the forward mapped reads - 0.5 is small,</span>
    <span class="c1"># this threhold is defined to be 0.1</span>
    <span class="k">if</span> <span class="n">np</span><span class="o">.</span><span class="n">isclose</span><span class="p">([</span><span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">forward_mapped_reads</span> <span class="o">-</span> <span class="mf">0.5</span><span class="p">)],</span> <span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">atol</span><span class="o">=</span><span class="mf">0.1</span><span class="p">):</span>
        <span class="k">return</span> <span class="s1">&#39;unstranded&#39;</span><span class="p">,</span> <span class="n">forward_mapped_reads</span><span class="p">,</span> <span class="n">reverse_mapped_reads</span><span class="p">,</span> <span class="n">total</span>
    <span class="k">elif</span> <span class="n">forward_mapped_reads</span> <span class="o">&gt;=</span> <span class="mf">0.5</span><span class="p">:</span>
        <span class="k">return</span> <span class="s1">&#39;forward&#39;</span><span class="p">,</span> <span class="n">forward_mapped_reads</span><span class="p">,</span> <span class="n">reverse_mapped_reads</span><span class="p">,</span> <span class="n">total</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">return</span> <span class="s1">&#39;reverse&#39;</span><span class="p">,</span> <span class="n">forward_mapped_reads</span><span class="p">,</span> <span class="n">reverse_mapped_reads</span><span class="p">,</span> <span class="n">total</span></div>
</pre></div>

          </div>
            
        </div>
        <div class="clearfix"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="../../index.html">riboraptor 0.3.0 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="../index.html" >Module code</a> &#187;</li> 
      </ul>
    </div>
<script type="text/javascript">
  $("#mobile-toggle a").click(function () {
    $("#left-column").toggle();
  });
</script>
<script type="text/javascript" src="../../_static/js/bootstrap.js"></script>
  <div class="footer">
    &copy; Copyright 2018, Saket Choudhary. Created using <a href="http://sphinx.pocoo.org/">Sphinx</a>.
  </div>
  </body>
</html>