
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="X-UA-Compatible" content="IE=Edge" />
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
  
  <!-- Licensed under the Apache 2.0 License -->
  <link rel="stylesheet" type="text/css" href="../../_static/fonts/open-sans/stylesheet.css" />
  <!-- Licensed under the SIL Open Font License -->
  <link rel="stylesheet" type="text/css" href="../../_static/fonts/source-serif-pro/source-serif-pro.css" />
  <link rel="stylesheet" type="text/css" href="../../_static/css/bootstrap.min.css" />
  <link rel="stylesheet" type="text/css" href="../../_static/css/bootstrap-theme.min.css" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
    <title>riboraptor.sequence &#8212; riboraptor 0.3.0 documentation</title>
    <link rel="stylesheet" href="../../_static/guzzle.css" type="text/css" />
    <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../../_static/graphviz.css" />
    <script type="text/javascript" id="documentation_options" data-url_root="../../" src="../../_static/documentation_options.js"></script>
    <script type="text/javascript" src="../../_static/jquery.js"></script>
    <script type="text/javascript" src="../../_static/underscore.js"></script>
    <script type="text/javascript" src="../../_static/doctools.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
  
   

  </head><body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="../../index.html">riboraptor 0.3.0 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="../index.html" accesskey="U">Module code</a> &#187;</li> 
      </ul>
    </div>
    <div class="container-wrapper">

      <div id="mobile-toggle">
        <a href="#"><span class="glyphicon glyphicon-align-justify" aria-hidden="true"></span></a>
      </div>
  <div id="left-column">
    <div class="sphinxsidebar">
<div class="sidebar-block">
  <div class="sidebar-wrapper">
    <h2>Contents</h2>
    <div class="sidebar-localtoc">
      <ul>
<li class="toctree-l1"><a class="reference internal" href="../../installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../example_workflow.html">Example Workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../cmd-manual.html">Manual</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../api-usage.html">API Usage</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../scores.html">Scores</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../modules.html">riboraptor</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../history.html">History</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../contributing.html">Contributing</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../authors.html">Credits</a></li>
</ul>

    </div>
  </div>
</div>
<div class="sidebar-block">
  <div class="sidebar-wrapper">
    <div id="main-search">
      <form class="form-inline" action="../../search.html" method="GET" role="form">
        <div class="input-group">
          <input name="q" type="text" class="form-control" placeholder="Search...">
        </div>
        <input type="hidden" name="check_keywords" value="yes" />
        <input type="hidden" name="area" value="default" />
      </form>
    </div>
  </div>
</div>
      
    </div>
  </div>
        <div id="right-column">
          
          <div role="navigation" aria-label="breadcrumbs navigation">
            <ol class="breadcrumb">
              <li><a href="../../index.html">Docs</a></li>
              
                <li><a href="../index.html">Module code</a></li>
              
              <li>riboraptor.sequence</li>
            </ol>
          </div>
          
          <div class="document clearer body">
            
  <h1>Source code for riboraptor.sequence</h1><div class="highlight"><pre>
<span></span><span class="sd">&quot;&quot;&quot;Utilities for extracting sequence from fasta.</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="kn">from</span> <span class="nn">__future__</span> <span class="k">import</span> <span class="n">absolute_import</span>
<span class="kn">from</span> <span class="nn">__future__</span> <span class="k">import</span> <span class="n">division</span>
<span class="kn">from</span> <span class="nn">__future__</span> <span class="k">import</span> <span class="n">print_function</span>
<span class="kn">from</span> <span class="nn">__future__</span> <span class="k">import</span> <span class="n">unicode_literals</span>

<span class="kn">import</span> <span class="nn">warnings</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">re</span>
<span class="kn">import</span> <span class="nn">sys</span>
<span class="kn">import</span> <span class="nn">pybedtools</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>

<span class="kn">from</span> <span class="nn">.genome</span> <span class="k">import</span> <span class="n">_get_sizes</span>
<span class="kn">from</span> <span class="nn">.genome</span> <span class="k">import</span> <span class="n">_get_bed</span>
<span class="kn">from</span> <span class="nn">.genome</span> <span class="k">import</span> <span class="n">__GENOMES_DB__</span>

<span class="kn">from</span> <span class="nn">.helpers</span> <span class="k">import</span> <span class="n">merge_intervals</span>

<span class="kn">from</span> <span class="nn">.interval</span> <span class="k">import</span> <span class="n">Interval</span>
<span class="kn">from</span> <span class="nn">.fasta</span> <span class="k">import</span> <span class="n">FastaReader</span>


<div class="viewcode-block" id="gene_sequence"><a class="viewcode-back" href="../../riboraptor.html#riboraptor.sequence.gene_sequence">[docs]</a><span class="k">def</span> <span class="nf">gene_sequence</span><span class="p">(</span><span class="n">gene_group</span><span class="p">,</span> <span class="n">fasta</span><span class="p">,</span> <span class="n">offset_5p</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">offset_3p</span><span class="o">=</span><span class="mi">0</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Extract seq genewise given coordinates in bed file</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    gene_group: DataFrame</span>
<span class="sd">                gene group from bed file</span>
<span class="sd">    fasta  str</span>
<span class="sd">           Path to fasta file</span>
<span class="sd">    offset_5p: int (positive)</span>
<span class="sd">               Number of bases to count upstream (5&#39;)</span>
<span class="sd">    offset_3p: int (positive)</span>
<span class="sd">               Number of bases to count downstream (3&#39;)</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    sequence: str</span>
<span class="sd">              sequence of the gene</span>
<span class="sd">    gene_offset_5p: Gene wise 5 prime offset</span>
<span class="sd">                    This might be different from `offset_5p` in cases where</span>
<span class="sd">                    `offset_5p` leads to a negative coordinate</span>
<span class="sd">    gene_offset_3p: Gene wise 3 prime offset</span>
<span class="sd">                    This might be different from `offset_3p` in cases where</span>
<span class="sd">                    `offset_3p` leads to position beyond chromsome length</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">offset_5p</span> <span class="o">&lt;</span> <span class="mi">0</span> <span class="ow">or</span> <span class="n">offset_3p</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span><span class="s1">&#39;Offsets must be non-negative&#39;</span><span class="p">)</span>
        <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">fasta</span><span class="p">,</span> <span class="n">FastaReader</span><span class="p">):</span>
        <span class="n">fasta</span> <span class="o">=</span> <span class="n">FastaReader</span><span class="p">(</span><span class="n">fasta</span><span class="p">)</span>
    <span class="n">chromosome_lengths</span> <span class="o">=</span> <span class="n">fasta</span><span class="o">.</span><span class="n">chromosomes</span>

    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">gene_group</span><span class="p">[</span><span class="s1">&#39;strand&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">())</span> <span class="o">!=</span> <span class="mi">1</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span><span class="s1">&#39;Multiple strands?: </span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">gene_group</span><span class="p">))</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">gene_group</span><span class="p">[</span><span class="s1">&#39;chrom&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">())</span> <span class="o">!=</span> <span class="mi">1</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span><span class="s1">&#39;Chromosome not unique for: </span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">gene_group</span><span class="p">))</span>
    <span class="n">chrom</span> <span class="o">=</span> <span class="n">gene_group</span><span class="p">[</span><span class="s1">&#39;chrom&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">strand</span> <span class="o">=</span> <span class="n">gene_group</span><span class="p">[</span><span class="s1">&#39;strand&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()[</span><span class="mi">0</span><span class="p">]</span>

    <span class="c1"># convert zero-based half-open interval to one-based full-closed</span>
    <span class="n">intervals</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span>
        <span class="nb">zip</span><span class="p">(</span><span class="n">gene_group</span><span class="p">[</span><span class="s1">&#39;chrom&#39;</span><span class="p">],</span> <span class="n">gene_group</span><span class="p">[</span><span class="s1">&#39;start&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span> <span class="n">gene_group</span><span class="p">[</span><span class="s1">&#39;end&#39;</span><span class="p">],</span>
            <span class="n">gene_group</span><span class="p">[</span><span class="s1">&#39;strand&#39;</span><span class="p">]))</span>

    <span class="n">intervals</span> <span class="o">=</span> <span class="p">[</span><span class="n">Interval</span><span class="p">(</span><span class="n">i</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">i</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">i</span><span class="p">[</span><span class="mi">2</span><span class="p">],</span> <span class="n">i</span><span class="p">[</span><span class="mi">3</span><span class="p">])</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">intervals</span><span class="p">]</span>
    <span class="n">intervals_combined</span><span class="p">,</span> <span class="n">gene_offset_5p</span><span class="p">,</span> <span class="n">gene_offset_3p</span> <span class="o">=</span> <span class="n">merge_intervals</span><span class="p">(</span>
        <span class="n">intervals</span><span class="p">,</span> <span class="n">chromosome_lengths</span><span class="p">,</span> <span class="n">offset_5p</span><span class="p">,</span> <span class="n">offset_3p</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>
    <span class="n">sequences</span> <span class="o">=</span> <span class="n">fasta</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="n">intervals_combined</span><span class="p">)</span>
    <span class="n">sequence</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">sequences</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">strand</span> <span class="o">==</span> <span class="s1">&#39;-&#39;</span><span class="p">:</span>
        <span class="n">sequence</span> <span class="o">=</span> <span class="n">fasta</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span>
    <span class="k">return</span> <span class="p">(</span><span class="n">sequence</span><span class="p">,</span> <span class="n">gene_offset_5p</span><span class="p">,</span> <span class="n">gene_offset_3p</span><span class="p">)</span></div>


<div class="viewcode-block" id="export_gene_sequences"><a class="viewcode-back" href="../../riboraptor.html#riboraptor.sequence.export_gene_sequences">[docs]</a><span class="k">def</span> <span class="nf">export_gene_sequences</span><span class="p">(</span><span class="n">bed</span><span class="p">,</span> <span class="n">fasta</span><span class="p">,</span> <span class="n">saveto</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">offset_5p</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">offset_3p</span><span class="o">=</span><span class="mi">0</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Export all gene sequences.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    bed: str</span>
<span class="sd">         Path to CDS or 5&#39;UTR or 3&#39;UTR bed</span>
<span class="sd">    fasta: str</span>
<span class="sd">           Path to fasta to fetch the sequences from</span>
<span class="sd">    saveto: str</span>
<span class="sd">            Path to write output tsv file</span>
<span class="sd">    offset_5p: int (positive)</span>
<span class="sd">               Number of bases to count upstream (5&#39;)</span>
<span class="sd">    offset_3p: int (positive)</span>
<span class="sd">               Number of bases to count downstream (3&#39;)</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    gene_profiles: file</span>
<span class="sd">                   with the following format:</span>
<span class="sd">                   gene1\t5poffset1\t3poffset1\tsequence\n</span>
<span class="sd">                   gene2\t5poffset2\t3poffset2\tsequence\n</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">bed</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;_&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="ow">in</span> <span class="n">__GENOMES_DB__</span><span class="p">:</span>
        <span class="n">genome</span><span class="p">,</span> <span class="n">region_type</span> <span class="o">=</span> <span class="n">bed</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;_&#39;</span><span class="p">)</span>
        <span class="n">bed</span> <span class="o">=</span> <span class="n">_get_bed</span><span class="p">(</span><span class="n">region_type</span><span class="p">,</span> <span class="n">genome</span><span class="p">)</span>
    <span class="n">bed_df</span> <span class="o">=</span> <span class="n">pybedtools</span><span class="o">.</span><span class="n">BedTool</span><span class="p">(</span><span class="n">bed</span><span class="p">)</span><span class="o">.</span><span class="n">sort</span><span class="p">()</span><span class="o">.</span><span class="n">to_dataframe</span><span class="p">()</span>
    <span class="n">bed_df</span><span class="p">[</span><span class="s1">&#39;chrom&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">bed_df</span><span class="p">[</span><span class="s1">&#39;chrom&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>
    <span class="n">bed_df</span><span class="p">[</span><span class="s1">&#39;name&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">bed_df</span><span class="p">[</span><span class="s1">&#39;name&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>
    <span class="n">bed_grouped</span> <span class="o">=</span> <span class="n">bed_df</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s1">&#39;name&#39;</span><span class="p">)</span>
    <span class="n">to_write</span> <span class="o">=</span> <span class="s1">&#39;gene_name</span><span class="se">\t</span><span class="s1">offset_5p</span><span class="se">\t</span><span class="s1">offset_3p</span><span class="se">\t</span><span class="s1">sequence</span><span class="se">\n</span><span class="s1">&#39;</span>
    <span class="k">for</span> <span class="n">gene_name</span><span class="p">,</span> <span class="n">gene_group</span> <span class="ow">in</span> <span class="n">bed_grouped</span><span class="p">:</span>
        <span class="n">sequence</span><span class="p">,</span> <span class="n">gene_offset_5p</span><span class="p">,</span> <span class="n">gene_offset_3p</span> <span class="o">=</span> <span class="n">gene_sequence</span><span class="p">(</span>
            <span class="n">gene_group</span><span class="p">,</span> <span class="n">fasta</span><span class="p">,</span> <span class="n">offset_5p</span><span class="p">,</span> <span class="n">offset_3p</span><span class="p">)</span>
        <span class="n">to_write</span> <span class="o">+=</span> <span class="s1">&#39;</span><span class="si">{}</span><span class="se">\t</span><span class="si">{}</span><span class="se">\t</span><span class="si">{}</span><span class="se">\t</span><span class="si">{}</span><span class="se">\n</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">gene_name</span><span class="p">,</span> <span class="nb">int</span><span class="p">(</span><span class="n">gene_offset_5p</span><span class="p">),</span>
                                              <span class="nb">int</span><span class="p">(</span><span class="n">gene_offset_3p</span><span class="p">),</span> <span class="n">sequence</span><span class="p">)</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">saveto</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">to_write</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">saveto</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">outfile</span><span class="p">:</span>
            <span class="n">outfile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">to_write</span><span class="p">)</span></div>
</pre></div>

          </div>
            
        </div>
        <div class="clearfix"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="../../index.html">riboraptor 0.3.0 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="../index.html" >Module code</a> &#187;</li> 
      </ul>
    </div>
<script type="text/javascript">
  $("#mobile-toggle a").click(function () {
    $("#left-column").toggle();
  });
</script>
<script type="text/javascript" src="../../_static/js/bootstrap.js"></script>
  <div class="footer">
    &copy; Copyright 2018, Saket Choudhary. Created using <a href="http://sphinx.pocoo.org/">Sphinx</a>.
  </div>
  </body>
</html>